<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Zebulun Arendsee" />


<title>Case Study: GFF processing</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
a.sourceLine { display: inline-block; min-height: 1.25em; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; }
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em; }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; color: #aaaaaa;  padding-left: 4px; }
@media screen {
a.sourceLine::before { text-decoration: underline; color: initial; }
}
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.bn { color: #40a070; } /* BaseN */
code span.fl { color: #40a070; } /* Float */
code span.ch { color: #4070a0; } /* Char */
code span.st { color: #4070a0; } /* String */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.ot { color: #007020; } /* Other */
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.fu { color: #06287e; } /* Function */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code span.cn { color: #880000; } /* Constant */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.ss { color: #bb6688; } /* SpecialString */
code span.im { } /* Import */
code span.va { color: #19177c; } /* Variable */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.op { color: #666666; } /* Operator */
code span.bu { } /* BuiltIn */
code span.ex { } /* Extension */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.at { color: #7d9029; } /* Attribute */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>

</head>

<body>




<h1 class="title toc-ignore">Case Study: GFF processing</h1>
<h4 class="author"><em>Zebulun Arendsee</em></h4>



<p>This work is funded by the National Science Foundation grant <a href="https://www.nsf.gov/awardsearch/showAward?AWD_ID=1546858">NSF-IOS 1546858</a>.</p>
<section id="the-problem" class="level2">
<h2>The problem</h2>
<p>Gene Feature Format (GFF) is used to annotate intervals on a genome. Loading and validating a GFF is a common first step in a bioinformatics analysis. Mistakes at this step can cause major problems later, so it is important to validate the GFF and report good diagnostics when problems arise. Here I will show how such a pipeline can be written using standard methods, and then show how <code>rmonad</code> can be used to organize and annotate such a pipeline.</p>
<p>GFF files are TAB-delimited where each row corresponds to a single interval. These intervals, though, may be ontologically related. For example a gene is the parent of an mRNA, which in turn is a parent of a set of exons and a set of coding sequences (CDS). These relations are specified in the attribute column (column 9). Here is an example (with TABs replaced with space) that introduces the main format and the variants we need to support:</p>
<pre><code># this is a comment, they can appear anywhere in the GFF
# also note, empty lines can appear anywhere in the file

# This is a simple mono-exonic gene
I   .   gene   11565   11951   .   -   .   ID=gene0
I   .   mRNA   11565   11951   .   -   .   ID=mrna0;Parent=gene0
I   .   exon   11565   11951   .   -   .   Parent=mrna0
I   .   CDS    11565   11951   .   -   0   Parent=mrna0

# this is a gene with two splicing variants:
I   .   gene   61931   83591   .   +   .   ID=gene1

I   .   mRNA   61931   83591   .   +   .   ID=rna1;Parent=gene1
I   .   exon   61931   62344   .   +   .   Parent=rna1
I   .   exon   81616   82209   .   +   .   Parent=rna1
I   .   exon   82211   83591   .   +   .   Parent=rna1
I   .   CDS   61931   62344   .   +   0   Parent=rna1
I   .   CDS   81616   82209   .   +   0   Parent=rna1
I   .   CDS   82211   82681   .   +   0   Parent=rna1

I   .   mRNA   61931   83591   .   +   .   ID=rna2;Parent=gene1
I   .   exon   61931   62344   .   +   .   Parent=rna2
I   .   exon   81616   82209   .   +   .   Parent=rna2
I   .   exon   82211   83591   .   +   .   Parent=rna2
I   .   CDS   61931   62344   .   +   0   Parent=rna2
I   .   CDS   82211   82681   .   +   0   Parent=rna2

# Below are a few variants that occur (unforunately) in the wild

# V1: CDS directly descending from a gene.
b   .   gene   7235   9016   .   -   .   ID=gene2
b   .   CDS    7235   9016   .   -   0   Parent=gene2

# V2: 'Parent=-' when feature has no parent
I   .   gene   11565   11951   .   -   .   ID=gene3;Parent=-

# V3: Elements with no tags that need to be treated as IDs
I   .   gene   11565   11951   .   -   .   gene3
I   .   mRNA   11565   11951   .   -   .   ID=mrna3;Parent=gene3
I   .   exon   11565   11951   .   -   .   Parent=mrna3
I   .   CDS    11565   11951   .   -   0   Parent=mrna3</code></pre>
<p>Another issues we need to account for is type synonyms. The feature type (column 3) is required to be valid Sequence Ontology (SO) terms. For the purposes of this vignette, I will just handle the following sets of synonyms:</p>
<ul>
<li><p>gene := <code>gene</code> | <code>SO:0000704</code></p></li>
<li><p>mRNA := <code>mRNA</code> | <code>messenger RNA</code> | <code>messenger_RNA</code> | <code>SO:0000234</code> | <code>transcript</code> | <code>SO:0000673</code></p></li>
<li><p>CDS := <code>CDS</code> | <code>coding_sequence</code> | <code>coding sequence</code> | <code>SO:0000316</code></p></li>
<li><p>exon := <code>exon</code> | <code>SO:0000147</code> | <code>coding_exon</code> | <code>coding exon</code> |<code>SO:0000195</code></p></li>
</ul>
<p>For mRNA and exon, I am merging to ontology terms (mRNA and transcript; exon and coding exon). Formally, this is incorrect, but pratically it is probably the right thing. Since these transformations may be wrong, they need to be noted.</p>
</section>
<section id="solution" class="level2">
<h2>Solution</h2>
<p>To test our solution, I use the <code>gff</code> rmonad dataset.</p>
<section id="conventional-approach" class="level3">
<h3>Conventional approach</h3>
<p>Before using <code>rmonad</code>, I will use a more conventional approach.</p>
<pre class="sourceCode r" id="cb2"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">library</span>(readr)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw">library</span>(dplyr)</a></code></pre>
<pre><code>## 
## Attaching package: 'dplyr'</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union</code></pre>
<pre class="sourceCode r" id="cb6"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">library</span>(tidyr)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw">library</span>(magrittr)</a></code></pre>
<pre><code>## 
## Attaching package: 'magrittr'</code></pre>
<pre><code>## The following object is masked from 'package:tidyr':
## 
##     extract</code></pre>
<pre class="sourceCode r" id="cb9"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="kw">library</span>(rmonad)</a></code></pre>
<pre><code>## 
## Attaching package: 'rmonad'</code></pre>
<pre><code>## The following object is masked from 'package:dplyr':
## 
##     combine</code></pre>
<pre class="sourceCode r" id="cb12"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw">set.seed</span>(<span class="dv">210</span>)</a>
<a class="sourceLine" id="cb12-2" data-line-number="2"></a>
<a class="sourceLine" id="cb12-3" data-line-number="3"><span class="kw">data</span>(gff)</a>
<a class="sourceLine" id="cb12-4" data-line-number="4"></a>
<a class="sourceLine" id="cb12-5" data-line-number="5">read_gff &lt;-<span class="st"> </span><span class="cf">function</span>(file){</a>
<a class="sourceLine" id="cb12-6" data-line-number="6">  readr<span class="op">::</span><span class="kw">read_tsv</span>(</a>
<a class="sourceLine" id="cb12-7" data-line-number="7">    file,</a>
<a class="sourceLine" id="cb12-8" data-line-number="8">    <span class="dt">col_names =</span> <span class="kw">c</span>(</a>
<a class="sourceLine" id="cb12-9" data-line-number="9">        <span class="st">&quot;seqid&quot;</span>,</a>
<a class="sourceLine" id="cb12-10" data-line-number="10">        <span class="st">&quot;source&quot;</span>,</a>
<a class="sourceLine" id="cb12-11" data-line-number="11">        <span class="st">&quot;type&quot;</span>,</a>
<a class="sourceLine" id="cb12-12" data-line-number="12">        <span class="st">&quot;start&quot;</span>,</a>
<a class="sourceLine" id="cb12-13" data-line-number="13">        <span class="st">&quot;stop&quot;</span>,</a>
<a class="sourceLine" id="cb12-14" data-line-number="14">        <span class="st">&quot;score&quot;</span>,</a>
<a class="sourceLine" id="cb12-15" data-line-number="15">        <span class="st">&quot;strand&quot;</span>,</a>
<a class="sourceLine" id="cb12-16" data-line-number="16">        <span class="st">&quot;phase&quot;</span>,</a>
<a class="sourceLine" id="cb12-17" data-line-number="17">        <span class="st">&quot;attr&quot;</span></a>
<a class="sourceLine" id="cb12-18" data-line-number="18">    ),</a>
<a class="sourceLine" id="cb12-19" data-line-number="19">    <span class="dt">na        =</span> <span class="st">&quot;.&quot;</span>,</a>
<a class="sourceLine" id="cb12-20" data-line-number="20">    <span class="dt">comment   =</span> <span class="st">&quot;#&quot;</span>,</a>
<a class="sourceLine" id="cb12-21" data-line-number="21">    <span class="dt">col_types =</span> <span class="st">&quot;ccciidcic&quot;</span></a>
<a class="sourceLine" id="cb12-22" data-line-number="22">  )</a>
<a class="sourceLine" id="cb12-23" data-line-number="23">}</a>
<a class="sourceLine" id="cb12-24" data-line-number="24"></a>
<a class="sourceLine" id="cb12-25" data-line-number="25"><span class="kw">read_gff</span>(gff<span class="op">$</span>good)</a></code></pre>
<pre><code>## # A tibble: 25 x 9
##    seqid source  type start  stop score strand phase                  attr
##    &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;dbl&gt;  &lt;chr&gt; &lt;int&gt;                 &lt;chr&gt;
##  1     I   &lt;NA&gt;  gene 11565 11951    NA      -    NA              ID=gene0
##  2     I   &lt;NA&gt;  mRNA 11565 11951    NA      -    NA ID=mrna0;Parent=gene0
##  3     I   &lt;NA&gt;  exon 11565 11951    NA      -    NA          Parent=mrna0
##  4     I   &lt;NA&gt;   CDS 11565 11951    NA      -     0          Parent=mrna0
##  5     I   &lt;NA&gt;  gene 61931 83591    NA      +    NA              ID=gene1
##  6     I   &lt;NA&gt;  mRNA 61931 83591    NA      +    NA  ID=rna1;Parent=gene1
##  7     I   &lt;NA&gt;  exon 61931 62344    NA      +    NA           Parent=rna1
##  8     I   &lt;NA&gt;  exon 81616 82209    NA      +    NA           Parent=rna1
##  9     I   &lt;NA&gt;  exon 82211 83591    NA      +    NA           Parent=rna1
## 10     I   &lt;NA&gt;   CDS 61931 62344    NA      +     0           Parent=rna1
## # ... with 15 more rows</code></pre>
<p>This is a nice start, also, <code>readr</code> will pick up on any deviations from the specified column number and type, warning of problems:</p>
<pre class="sourceCode r" id="cb14"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw">read_gff</span>(gff<span class="op">$</span>invalid_type)</a></code></pre>
<pre><code>## Warning in rbind(names(probs), probs_f): number of columns of result is not
## a multiple of vector length (arg 2)</code></pre>
<pre><code>## Warning: 1 parsing failure.
## row # A tibble: 1 x 5 col     row   col   expected                        actual         file expected   &lt;int&gt; &lt;chr&gt;      &lt;chr&gt;                         &lt;chr&gt;        &lt;chr&gt; actual 1     1 start an integer this is not an integer at all literal data file # A tibble: 1 x 5</code></pre>
<pre><code>## # A tibble: 2 x 9
##   seqid source  type start  stop score strand phase                  attr
##   &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;dbl&gt;  &lt;chr&gt; &lt;int&gt;                 &lt;chr&gt;
## 1     I   &lt;NA&gt;  gene    NA 11951    NA      -    NA              ID=gene0
## 2     I   &lt;NA&gt;  mRNA 11565 11951    NA      -    NA ID=mrna0;Parent=gene0</code></pre>
<pre class="sourceCode r" id="cb18"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"><span class="kw">read_gff</span>(gff<span class="op">$</span>not_a_gff1)</a></code></pre>
<pre><code>## Warning in rbind(names(probs), probs_f): number of columns of result is not
## a multiple of vector length (arg 1)</code></pre>
<pre><code>## Warning: 3 parsing failures.
## row # A tibble: 3 x 5 col     row   col  expected    actual         file expected   &lt;int&gt; &lt;chr&gt;     &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt; actual 1     1  &lt;NA&gt; 9 columns 1 columns literal data file 2     2  &lt;NA&gt; 9 columns 1 columns literal data row 3     3  &lt;NA&gt; 9 columns 1 columns literal data</code></pre>
<pre><code>## # A tibble: 3 x 9
##                seqid source  type start  stop score strand phase  attr
##                &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;dbl&gt;  &lt;chr&gt; &lt;int&gt; &lt;chr&gt;
## 1   Once upon a time   &lt;NA&gt;  &lt;NA&gt;    NA    NA    NA   &lt;NA&gt;    NA  &lt;NA&gt;
## 2 there was a prince   &lt;NA&gt;  &lt;NA&gt;    NA    NA    NA   &lt;NA&gt;    NA  &lt;NA&gt;
## 3 who liked to plumb   &lt;NA&gt;  &lt;NA&gt;    NA    NA    NA   &lt;NA&gt;    NA  &lt;NA&gt;</code></pre>
<p>R uses <code>NA</code> to indicate missing values. The R ‘numeric’ type corresponds to Haskell <code>[Maybe Num]</code>, i.e. an array of possibly empty values. In a GFF, columns 2,6,7,8 and 9 may be missing, the others may not. So we need an additional assertion that these are complete.</p>
<pre class="sourceCode r" id="cb22"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1">g &lt;-<span class="st"> </span><span class="kw">read_gff</span>(gff<span class="op">$</span>good)</a>
<a class="sourceLine" id="cb22-2" data-line-number="2"><span class="cf">for</span>(col <span class="cf">in</span> <span class="kw">c</span>(<span class="st">&quot;seqid&quot;</span>, <span class="st">&quot;type&quot;</span>, <span class="st">&quot;start&quot;</span>, <span class="st">&quot;stop&quot;</span>)){</a>
<a class="sourceLine" id="cb22-3" data-line-number="3">  <span class="cf">if</span>(<span class="kw">any</span>(<span class="kw">is.na</span>(g[[col]]))){</a>
<a class="sourceLine" id="cb22-4" data-line-number="4">    <span class="kw">stop</span>(<span class="st">&quot;GFFError: Column '&quot;</span>, col, <span class="st">&quot;' may not have missing values&quot;</span>)</a>
<a class="sourceLine" id="cb22-5" data-line-number="5">  }</a>
<a class="sourceLine" id="cb22-6" data-line-number="6">}</a></code></pre>
<p>Now we need to account for type synonyms</p>
<pre class="sourceCode r" id="cb23"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1">gene_synonyms &lt;-<span class="st"> 'SO:0000704'</span></a>
<a class="sourceLine" id="cb23-2" data-line-number="2">mRNA_synonyms &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">'messenger_RNA'</span>, <span class="st">'messenger RNA'</span>, <span class="st">'SO:0000234'</span>)</a>
<a class="sourceLine" id="cb23-3" data-line-number="3">CDS_synonyms  &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">'coding_sequence'</span>, <span class="st">'coding sequence'</span>, <span class="st">'SO:0000316'</span>)</a>
<a class="sourceLine" id="cb23-4" data-line-number="4">exon_synonyms &lt;-<span class="st"> 'SO:0000147'</span></a>
<a class="sourceLine" id="cb23-5" data-line-number="5"></a>
<a class="sourceLine" id="cb23-6" data-line-number="6">g<span class="op">$</span>type &lt;-<span class="st"> </span><span class="kw">ifelse</span>(g<span class="op">$</span>type <span class="op">%in%</span><span class="st"> </span>gene_synonyms, <span class="st">'gene'</span>, g<span class="op">$</span>type)</a>
<a class="sourceLine" id="cb23-7" data-line-number="7">g<span class="op">$</span>type &lt;-<span class="st"> </span><span class="kw">ifelse</span>(g<span class="op">$</span>type <span class="op">%in%</span><span class="st"> </span>mRNA_synonyms, <span class="st">'mRNA'</span>, g<span class="op">$</span>type)</a>
<a class="sourceLine" id="cb23-8" data-line-number="8">g<span class="op">$</span>type &lt;-<span class="st"> </span><span class="kw">ifelse</span>(g<span class="op">$</span>type <span class="op">%in%</span><span class="st"> </span>CDS_synonyms,  <span class="st">'CDS'</span>,  g<span class="op">$</span>type)</a>
<a class="sourceLine" id="cb23-9" data-line-number="9">g<span class="op">$</span>type &lt;-<span class="st"> </span><span class="kw">ifelse</span>(g<span class="op">$</span>type <span class="op">%in%</span><span class="st"> </span>exon_synonyms, <span class="st">'exon'</span>, g<span class="op">$</span>type)</a>
<a class="sourceLine" id="cb23-10" data-line-number="10"></a>
<a class="sourceLine" id="cb23-11" data-line-number="11">mRNA_near_synonyms &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">'transcript'</span>, <span class="st">'SO:0000673'</span>)</a>
<a class="sourceLine" id="cb23-12" data-line-number="12">exon_near_synonyms &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">'SO:0000147'</span>, <span class="st">'coding_exon'</span>, <span class="st">'coding exon'</span>, <span class="st">'SO:0000195'</span>)</a>
<a class="sourceLine" id="cb23-13" data-line-number="13"></a>
<a class="sourceLine" id="cb23-14" data-line-number="14"><span class="cf">if</span>(<span class="kw">any</span>(g<span class="op">$</span>type <span class="op">%in%</span><span class="st"> </span>mRNA_near_synonyms)){</a>
<a class="sourceLine" id="cb23-15" data-line-number="15">    g<span class="op">$</span>type &lt;-<span class="st"> </span><span class="kw">ifelse</span>(g<span class="op">$</span>type <span class="op">%in%</span><span class="st"> </span>mRNA_near_synonyms, <span class="st">'mRNA'</span>, g<span class="op">$</span>type)</a>
<a class="sourceLine" id="cb23-16" data-line-number="16">    <span class="kw">warning</span>(<span class="st">&quot;Substituting transcript types for mRNA types, this is probably OK&quot;</span>)</a>
<a class="sourceLine" id="cb23-17" data-line-number="17">}</a>
<a class="sourceLine" id="cb23-18" data-line-number="18"></a>
<a class="sourceLine" id="cb23-19" data-line-number="19"><span class="cf">if</span>(<span class="kw">any</span>(g<span class="op">$</span>type <span class="op">%in%</span><span class="st"> </span>exon_near_synonyms)){</a>
<a class="sourceLine" id="cb23-20" data-line-number="20">    g<span class="op">$</span>type &lt;-<span class="st"> </span><span class="kw">ifelse</span>(g<span class="op">$</span>type <span class="op">%in%</span><span class="st"> </span>exon_near_synonyms, <span class="st">'exon'</span>, g<span class="op">$</span>type)</a>
<a class="sourceLine" id="cb23-21" data-line-number="21">    <span class="kw">warning</span>(<span class="st">&quot;Substituting transcript types for exon types, this is probably OK&quot;</span>)</a>
<a class="sourceLine" id="cb23-22" data-line-number="22">}</a></code></pre>
<p>Now we need to evaluate the attribute column.</p>
<pre class="sourceCode r" id="cb24"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1">tags &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;ID&quot;</span>, <span class="st">&quot;Parent&quot;</span>)</a>
<a class="sourceLine" id="cb24-2" data-line-number="2"></a>
<a class="sourceLine" id="cb24-3" data-line-number="3"><span class="kw">data_frame</span>(</a>
<a class="sourceLine" id="cb24-4" data-line-number="4">    <span class="dt">attr  =</span> stringr<span class="op">::</span><span class="kw">str_split</span>(g<span class="op">$</span>attr, <span class="st">&quot;;&quot;</span>),</a>
<a class="sourceLine" id="cb24-5" data-line-number="5">    <span class="dt">order =</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(g)</a>
<a class="sourceLine" id="cb24-6" data-line-number="6">  ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb24-7" data-line-number="7"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">mutate</span>(<span class="dt">ntags =</span> <span class="kw">sapply</span>(attr, length)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb24-8" data-line-number="8"><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">unnest</span>(attr) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb24-9" data-line-number="9"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">mutate</span>(<span class="dt">attr =</span> <span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">'='</span>, attr), attr, <span class="kw">paste</span>(<span class="st">&quot;.U&quot;</span>, attr, <span class="dt">sep=</span><span class="st">&quot;=&quot;</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb24-10" data-line-number="10"><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">separate_</span>(</a>
<a class="sourceLine" id="cb24-11" data-line-number="11">    <span class="dt">col   =</span> <span class="st">&quot;attr&quot;</span>,</a>
<a class="sourceLine" id="cb24-12" data-line-number="12">    <span class="dt">into  =</span> <span class="kw">c</span>(<span class="st">&quot;tag&quot;</span>, <span class="st">&quot;value&quot;</span>),</a>
<a class="sourceLine" id="cb24-13" data-line-number="13">    <span class="dt">sep   =</span> <span class="st">&quot;=&quot;</span>,</a>
<a class="sourceLine" id="cb24-14" data-line-number="14">    <span class="dt">extra =</span> <span class="st">&quot;merge&quot;</span></a>
<a class="sourceLine" id="cb24-15" data-line-number="15">  ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb24-16" data-line-number="16"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(tag <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(tags, <span class="st">&quot;.U&quot;</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb24-17" data-line-number="17"><span class="st">  </span>{</a>
<a class="sourceLine" id="cb24-18" data-line-number="18">    <span class="cf">if</span>(<span class="kw">nrow</span>(.) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>){</a>
<a class="sourceLine" id="cb24-19" data-line-number="19">      tidyr<span class="op">::</span><span class="kw">spread</span>(., <span class="dt">key=</span><span class="st">&quot;tag&quot;</span>, <span class="dt">value=</span><span class="st">&quot;value&quot;</span>)</a>
<a class="sourceLine" id="cb24-20" data-line-number="20">    } <span class="cf">else</span> {</a>
<a class="sourceLine" id="cb24-21" data-line-number="21">      .<span class="op">$</span>tag   =<span class="st"> </span><span class="ot">NULL</span></a>
<a class="sourceLine" id="cb24-22" data-line-number="22">      .<span class="op">$</span>value =<span class="st"> </span><span class="ot">NULL</span></a>
<a class="sourceLine" id="cb24-23" data-line-number="23">      .</a>
<a class="sourceLine" id="cb24-24" data-line-number="24">    }</a>
<a class="sourceLine" id="cb24-25" data-line-number="25">  } <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb24-26" data-line-number="26"><span class="st">  </span>{</a>
<a class="sourceLine" id="cb24-27" data-line-number="27">    <span class="cf">if</span>(<span class="st">&quot;Parent&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">names</span>(.)){</a>
<a class="sourceLine" id="cb24-28" data-line-number="28">      .<span class="op">$</span>Parent &lt;-<span class="st"> </span><span class="kw">ifelse</span>(.<span class="op">$</span>Parent <span class="op">==</span><span class="st"> &quot;-&quot;</span>, <span class="ot">NA</span>, .<span class="op">$</span>Parent)</a>
<a class="sourceLine" id="cb24-29" data-line-number="29">    }</a>
<a class="sourceLine" id="cb24-30" data-line-number="30">    .</a>
<a class="sourceLine" id="cb24-31" data-line-number="31">  } <span class="op">%&gt;%</span><span class="st"> </span>{</a>
<a class="sourceLine" id="cb24-32" data-line-number="32">    <span class="cf">for</span>(tag <span class="cf">in</span> <span class="kw">c</span>(tags, <span class="st">&quot;.U&quot;</span>)){</a>
<a class="sourceLine" id="cb24-33" data-line-number="33">      <span class="cf">if</span>(<span class="op">!</span><span class="st"> </span>tag <span class="op">%in%</span><span class="st"> </span><span class="kw">names</span>(.))</a>
<a class="sourceLine" id="cb24-34" data-line-number="34">        .[[tag]] =<span class="st"> </span><span class="ot">NA_character_</span></a>
<a class="sourceLine" id="cb24-35" data-line-number="35">    }</a>
<a class="sourceLine" id="cb24-36" data-line-number="36">    .</a>
<a class="sourceLine" id="cb24-37" data-line-number="37">  } <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb24-38" data-line-number="38"><span class="st">  </span>{</a>
<a class="sourceLine" id="cb24-39" data-line-number="39">    <span class="cf">if</span>(<span class="st">&quot;ID&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">names</span>(.))</a>
<a class="sourceLine" id="cb24-40" data-line-number="40">      .<span class="op">$</span>ID &lt;-<span class="st"> </span><span class="kw">ifelse</span>(<span class="kw">is.na</span>(.<span class="op">$</span>ID) <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(.<span class="op">$</span>.U) <span class="op">&amp;</span><span class="st"> </span>.<span class="op">$</span>ntags <span class="op">==</span><span class="st"> </span><span class="dv">1</span>, .<span class="op">$</span>.U, .<span class="op">$</span>ID)</a>
<a class="sourceLine" id="cb24-41" data-line-number="41">    .</a>
<a class="sourceLine" id="cb24-42" data-line-number="42">  } <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb24-43" data-line-number="43"><span class="st">  </span><span class="kw">merge</span>(<span class="kw">data_frame</span>(<span class="dt">order=</span><span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(g)), <span class="dt">all=</span><span class="ot">TRUE</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb24-44" data-line-number="44"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">arrange</span>(order) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb24-45" data-line-number="45"><span class="st">  </span>{ <span class="kw">cbind</span>(g, .) } <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb24-46" data-line-number="46"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">select</span>(<span class="op">-</span>.U, <span class="op">-</span>order, <span class="op">-</span>ntags, <span class="op">-</span>attr) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb24-47" data-line-number="47"><span class="st">  </span>{</a>
<a class="sourceLine" id="cb24-48" data-line-number="48">    <span class="cf">if</span>(<span class="kw">all</span>(<span class="kw">c</span>(<span class="st">&quot;ID&quot;</span>, <span class="st">&quot;Parent&quot;</span>) <span class="op">%in%</span><span class="st"> </span><span class="kw">names</span>(.))){</a>
<a class="sourceLine" id="cb24-49" data-line-number="49">      parents &lt;-<span class="st"> </span><span class="kw">subset</span>(., type <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;CDS&quot;</span>, <span class="st">&quot;exon&quot;</span>))<span class="op">$</span>Parent</a>
<a class="sourceLine" id="cb24-50" data-line-number="50">      parent_types &lt;-<span class="st"> </span><span class="kw">subset</span>(., ID <span class="op">%in%</span><span class="st"> </span>parents)<span class="op">$</span>type</a>
<a class="sourceLine" id="cb24-51" data-line-number="51"></a>
<a class="sourceLine" id="cb24-52" data-line-number="52">      <span class="cf">if</span>(<span class="kw">any</span>(parent_types <span class="op">==</span><span class="st"> &quot;gene&quot;</span>))</a>
<a class="sourceLine" id="cb24-53" data-line-number="53">        <span class="kw">warning</span>(<span class="st">&quot;Found CDS or exon directly inheriting from a gene, this may be fine.&quot;</span>) </a>
<a class="sourceLine" id="cb24-54" data-line-number="54"></a>
<a class="sourceLine" id="cb24-55" data-line-number="55">      <span class="cf">if</span>(<span class="op">!</span><span class="st"> </span><span class="kw">all</span>(parent_types <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;gene&quot;</span>, <span class="st">&quot;mRNA&quot;</span>)))</a>
<a class="sourceLine" id="cb24-56" data-line-number="56">        <span class="kw">stop</span>(<span class="st">&quot;Found CDS or exon with illegal parent&quot;</span>)</a>
<a class="sourceLine" id="cb24-57" data-line-number="57"></a>
<a class="sourceLine" id="cb24-58" data-line-number="58">      <span class="cf">if</span>( <span class="kw">any</span>(<span class="kw">is.na</span>(parents)) )</a>
<a class="sourceLine" id="cb24-59" data-line-number="59">        <span class="kw">stop</span>(<span class="st">&quot;Found CDS or exon with no parent&quot;</span>)</a>
<a class="sourceLine" id="cb24-60" data-line-number="60"></a>
<a class="sourceLine" id="cb24-61" data-line-number="61">      <span class="cf">if</span>(<span class="op">!</span><span class="st"> </span><span class="kw">any</span>(<span class="kw">duplicated</span>(.<span class="op">$</span>ID, <span class="dt">incomparables=</span><span class="ot">NA</span>)))</a>
<a class="sourceLine" id="cb24-62" data-line-number="62">        <span class="kw">warning</span>(<span class="st">&quot;IDs are not unique, this is probably bad&quot;</span>)</a>
<a class="sourceLine" id="cb24-63" data-line-number="63">    }</a>
<a class="sourceLine" id="cb24-64" data-line-number="64">    .</a>
<a class="sourceLine" id="cb24-65" data-line-number="65">  }</a></code></pre>
<pre><code>## Warning in function_list[[k]](value): Found CDS or exon directly inheriting
## from a gene, this may be fine.</code></pre>
<pre><code>##    seqid source type start  stop score strand phase    ID Parent
## 1      I   &lt;NA&gt; gene 11565 11951    NA      -    NA gene0   &lt;NA&gt;
## 2      I   &lt;NA&gt; mRNA 11565 11951    NA      -    NA mrna0  gene0
## 3      I   &lt;NA&gt; exon 11565 11951    NA      -    NA  &lt;NA&gt;  mrna0
## 4      I   &lt;NA&gt;  CDS 11565 11951    NA      -     0  &lt;NA&gt;  mrna0
## 5      I   &lt;NA&gt; gene 61931 83591    NA      +    NA gene1   &lt;NA&gt;
## 6      I   &lt;NA&gt; mRNA 61931 83591    NA      +    NA  rna1  gene1
## 7      I   &lt;NA&gt; exon 61931 62344    NA      +    NA  &lt;NA&gt;   rna1
## 8      I   &lt;NA&gt; exon 81616 82209    NA      +    NA  &lt;NA&gt;   rna1
## 9      I   &lt;NA&gt; exon 82211 83591    NA      +    NA  &lt;NA&gt;   rna1
## 10     I   &lt;NA&gt;  CDS 61931 62344    NA      +     0  &lt;NA&gt;   rna1
## 11     I   &lt;NA&gt;  CDS 81616 82209    NA      +     0  &lt;NA&gt;   rna1
## 12     I   &lt;NA&gt;  CDS 82211 82681    NA      +     0  &lt;NA&gt;   rna1
## 13     I   &lt;NA&gt; mRNA 61931 83591    NA      +    NA  rna2  gene1
## 14     I   &lt;NA&gt; exon 61931 62344    NA      +    NA  &lt;NA&gt;   rna2
## 15     I   &lt;NA&gt; exon 81616 82209    NA      +    NA  &lt;NA&gt;   rna2
## 16     I   &lt;NA&gt; exon 82211 83591    NA      +    NA  &lt;NA&gt;   rna2
## 17     I   &lt;NA&gt;  CDS 61931 62344    NA      +     0  &lt;NA&gt;   rna2
## 18     I   &lt;NA&gt;  CDS 82211 82681    NA      +     0  &lt;NA&gt;   rna2
## 19     b   &lt;NA&gt; gene  7235  9016    NA      -    NA gene2   &lt;NA&gt;
## 20     b   &lt;NA&gt;  CDS  7235  9016    NA      -     0  &lt;NA&gt;  gene2
## 21     I   &lt;NA&gt; gene 11565 11951    NA      -    NA gene3   &lt;NA&gt;
## 22     I   &lt;NA&gt; gene 11565 11951    NA      -    NA gene3   &lt;NA&gt;
## 23     I   &lt;NA&gt; mRNA 11565 11951    NA      -    NA mrna3  gene3
## 24     I   &lt;NA&gt; exon 11565 11951    NA      -    NA  &lt;NA&gt;  mrna3
## 25     I   &lt;NA&gt;  CDS 11565 11951    NA      -     0  &lt;NA&gt;  mrna3</code></pre>
<p>The beauty of this chain is that it requires few temporary variables (just <code>g</code>, and <code>tags</code>), it is a pure flow of data. It is an elegant sequence of functions operating on a single thread of data.</p>
<p>But there are a few problems.</p>
<p>First, it is in dire need of documentation. We could add in comments. But comments cannot be formatted well. A better approach is some form of literate programming, such as rewriting the program in Rmarkdown. But this 1) breaks the pipeline (since we can’t pipe between chunks), 2) results in an object we can’t compute on, 3) makes debugging even more difficult, because our code is spread out.</p>
</section>
<section id="rmonad-approach" class="level3">
<h3><code>rmonad</code> approach</h3>
<p>With <code>rmonad</code> we can mingle documentation and code in a computable object.</p>
<pre class="sourceCode r" id="cb27"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1">read_gff &lt;-<span class="st"> </span><span class="cf">function</span>(file, tags){</a>
<a class="sourceLine" id="cb27-2" data-line-number="2"></a>
<a class="sourceLine" id="cb27-3" data-line-number="3">  raw_gff &lt;-<span class="st"> </span><span class="kw">as_monad</span>(</a>
<a class="sourceLine" id="cb27-4" data-line-number="4">    {</a>
<a class="sourceLine" id="cb27-5" data-line-number="5"></a>
<a class="sourceLine" id="cb27-6" data-line-number="6">      <span class="st">&quot;</span></a>
<a class="sourceLine" id="cb27-7" data-line-number="7"><span class="st">      Rmonad supports docstrings. If an block begins with a string, this</span></a>
<a class="sourceLine" id="cb27-8" data-line-number="8"><span class="st">      string is extracted and stored. Python has something similar, where the</span></a>
<a class="sourceLine" id="cb27-9" data-line-number="9"><span class="st">      first string in a function is cast as documentation.</span></a>
<a class="sourceLine" id="cb27-10" data-line-number="10"><span class="st">      </span></a>
<a class="sourceLine" id="cb27-11" data-line-number="11"><span class="st">      The `as_monad` function takes an expression and wraps its result into a</span></a>
<a class="sourceLine" id="cb27-12" data-line-number="12"><span class="st">      context. It also handles the extraction of this docstring. The result</span></a>
<a class="sourceLine" id="cb27-13" data-line-number="13"><span class="st">      here is used at more than one place in the pipeline. Rather than</span></a>
<a class="sourceLine" id="cb27-14" data-line-number="14"><span class="st">      accessing it later as a global, it will be funneled bach in.</span></a>
<a class="sourceLine" id="cb27-15" data-line-number="15"><span class="st">      &quot;</span></a>
<a class="sourceLine" id="cb27-16" data-line-number="16"></a>
<a class="sourceLine" id="cb27-17" data-line-number="17">      readr<span class="op">::</span><span class="kw">read_tsv</span>(</a>
<a class="sourceLine" id="cb27-18" data-line-number="18">        file,</a>
<a class="sourceLine" id="cb27-19" data-line-number="19">        <span class="dt">col_names =</span> <span class="kw">c</span>(</a>
<a class="sourceLine" id="cb27-20" data-line-number="20">          <span class="st">&quot;seqid&quot;</span>,</a>
<a class="sourceLine" id="cb27-21" data-line-number="21">          <span class="st">&quot;source&quot;</span>,</a>
<a class="sourceLine" id="cb27-22" data-line-number="22">          <span class="st">&quot;type&quot;</span>,</a>
<a class="sourceLine" id="cb27-23" data-line-number="23">          <span class="st">&quot;start&quot;</span>,</a>
<a class="sourceLine" id="cb27-24" data-line-number="24">          <span class="st">&quot;stop&quot;</span>,</a>
<a class="sourceLine" id="cb27-25" data-line-number="25">          <span class="st">&quot;score&quot;</span>,</a>
<a class="sourceLine" id="cb27-26" data-line-number="26">          <span class="st">&quot;strand&quot;</span>,</a>
<a class="sourceLine" id="cb27-27" data-line-number="27">          <span class="st">&quot;phase&quot;</span>,</a>
<a class="sourceLine" id="cb27-28" data-line-number="28">          <span class="st">&quot;attr&quot;</span></a>
<a class="sourceLine" id="cb27-29" data-line-number="29">        ),</a>
<a class="sourceLine" id="cb27-30" data-line-number="30">        <span class="dt">na        =</span> <span class="st">&quot;.&quot;</span>,</a>
<a class="sourceLine" id="cb27-31" data-line-number="31">        <span class="dt">comment   =</span> <span class="st">&quot;#&quot;</span>,</a>
<a class="sourceLine" id="cb27-32" data-line-number="32">        <span class="dt">col_types =</span> <span class="st">&quot;ccciidcic&quot;</span></a>
<a class="sourceLine" id="cb27-33" data-line-number="33">      )</a>
<a class="sourceLine" id="cb27-34" data-line-number="34">    }</a>
<a class="sourceLine" id="cb27-35" data-line-number="35">  )</a>
<a class="sourceLine" id="cb27-36" data-line-number="36"></a>
<a class="sourceLine" id="cb27-37" data-line-number="37">  raw_gff <span class="op">%&gt;&gt;%</span><span class="st"> </span>{</a>
<a class="sourceLine" id="cb27-38" data-line-number="38"></a>
<a class="sourceLine" id="cb27-39" data-line-number="39">    <span class="st">&quot;</span></a>
<a class="sourceLine" id="cb27-40" data-line-number="40"><span class="st">    The %&gt;&gt;% operator applies the function described in this block to the</span></a>
<a class="sourceLine" id="cb27-41" data-line-number="41"><span class="st">    input on the left-hand-side. This corresponds to the UNIX '|' or magrittr's</span></a>
<a class="sourceLine" id="cb27-42" data-line-number="42"><span class="st">    '%&gt;%'. It differs from them in that it is a monadic bind operator, rather</span></a>
<a class="sourceLine" id="cb27-43" data-line-number="43"><span class="st">    than an application operator. It carries a context along with the</span></a>
<a class="sourceLine" id="cb27-44" data-line-number="44"><span class="st">    computations. The context can store past values, performance information,</span></a>
<a class="sourceLine" id="cb27-45" data-line-number="45"><span class="st">    this docstring, and links to the parent chunk. The context is a directed</span></a>
<a class="sourceLine" id="cb27-46" data-line-number="46"><span class="st">    graph of code chunks and their metadata.</span></a>
<a class="sourceLine" id="cb27-47" data-line-number="47"><span class="st">    &quot;</span></a>
<a class="sourceLine" id="cb27-48" data-line-number="48"></a>
<a class="sourceLine" id="cb27-49" data-line-number="49">    <span class="cf">for</span>(col <span class="cf">in</span> <span class="kw">c</span>(<span class="st">&quot;seqid&quot;</span>, <span class="st">&quot;type&quot;</span>, <span class="st">&quot;start&quot;</span>, <span class="st">&quot;stop&quot;</span>)){</a>
<a class="sourceLine" id="cb27-50" data-line-number="50">      <span class="cf">if</span>(<span class="kw">any</span>(<span class="kw">is.na</span>(.[[col]]))){</a>
<a class="sourceLine" id="cb27-51" data-line-number="51">        <span class="kw">stop</span>(<span class="st">&quot;GFFError: Column '&quot;</span>, col, <span class="st">&quot;' may not have missing values&quot;</span>)</a>
<a class="sourceLine" id="cb27-52" data-line-number="52">      }</a>
<a class="sourceLine" id="cb27-53" data-line-number="53">    }</a>
<a class="sourceLine" id="cb27-54" data-line-number="54">    .</a>
<a class="sourceLine" id="cb27-55" data-line-number="55">  } <span class="op">%&gt;&gt;%</span><span class="st"> </span>{</a>
<a class="sourceLine" id="cb27-56" data-line-number="56"></a>
<a class="sourceLine" id="cb27-57" data-line-number="57">    <span class="st">&quot;</span></a>
<a class="sourceLine" id="cb27-58" data-line-number="58"><span class="st">    Note that these blocks of code are copied verbatim from above, only using</span></a>
<a class="sourceLine" id="cb27-59" data-line-number="59"><span class="st">    '.' in place of 'g'.</span></a>
<a class="sourceLine" id="cb27-60" data-line-number="60"><span class="st">    &quot;</span></a>
<a class="sourceLine" id="cb27-61" data-line-number="61"></a>
<a class="sourceLine" id="cb27-62" data-line-number="62">    gene_synonyms &lt;-<span class="st"> 'SO:0000704'</span></a>
<a class="sourceLine" id="cb27-63" data-line-number="63">    mRNA_synonyms &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">'messenger_RNA'</span>, <span class="st">'messenger RNA'</span>, <span class="st">'SO:0000234'</span>)</a>
<a class="sourceLine" id="cb27-64" data-line-number="64">    CDS_synonyms  &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">'coding_sequence'</span>, <span class="st">'coding sequence'</span>, <span class="st">'SO:0000316'</span>)</a>
<a class="sourceLine" id="cb27-65" data-line-number="65">    exon_synonyms &lt;-<span class="st"> 'SO:0000147'</span></a>
<a class="sourceLine" id="cb27-66" data-line-number="66"></a>
<a class="sourceLine" id="cb27-67" data-line-number="67">    .<span class="op">$</span>type &lt;-<span class="st"> </span><span class="kw">ifelse</span>(.<span class="op">$</span>type <span class="op">%in%</span><span class="st"> </span>gene_synonyms, <span class="st">'gene'</span>, .<span class="op">$</span>type)</a>
<a class="sourceLine" id="cb27-68" data-line-number="68">    .<span class="op">$</span>type &lt;-<span class="st"> </span><span class="kw">ifelse</span>(.<span class="op">$</span>type <span class="op">%in%</span><span class="st"> </span>mRNA_synonyms, <span class="st">'mRNA'</span>, .<span class="op">$</span>type)</a>
<a class="sourceLine" id="cb27-69" data-line-number="69">    .<span class="op">$</span>type &lt;-<span class="st"> </span><span class="kw">ifelse</span>(.<span class="op">$</span>type <span class="op">%in%</span><span class="st"> </span>CDS_synonyms,  <span class="st">'CDS'</span>,  .<span class="op">$</span>type)</a>
<a class="sourceLine" id="cb27-70" data-line-number="70">    .<span class="op">$</span>type &lt;-<span class="st"> </span><span class="kw">ifelse</span>(.<span class="op">$</span>type <span class="op">%in%</span><span class="st"> </span>exon_synonyms, <span class="st">'exon'</span>, .<span class="op">$</span>type)</a>
<a class="sourceLine" id="cb27-71" data-line-number="71"></a>
<a class="sourceLine" id="cb27-72" data-line-number="72">    .</a>
<a class="sourceLine" id="cb27-73" data-line-number="73">  } <span class="op">%&gt;_%</span><span class="st"> </span>{</a>
<a class="sourceLine" id="cb27-74" data-line-number="74"></a>
<a class="sourceLine" id="cb27-75" data-line-number="75">    <span class="st">&quot;</span></a>
<a class="sourceLine" id="cb27-76" data-line-number="76"><span class="st">    The %&gt;_% operator lets this chunk of code be run for its effects, which</span></a>
<a class="sourceLine" id="cb27-77" data-line-number="77"><span class="st">    are emitting warnings if we replace the type with a questionable synonym.</span></a>
<a class="sourceLine" id="cb27-78" data-line-number="78"><span class="st">    We could alternatively just use %&gt;&gt;% and add a terminal '.' to this chunk.</span></a>
<a class="sourceLine" id="cb27-79" data-line-number="79"><span class="st">    The use of this operator, though, signals an interdependent branch. Where</span></a>
<a class="sourceLine" id="cb27-80" data-line-number="80"><span class="st">    failure of this branch triggers failure downstream.</span></a>
<a class="sourceLine" id="cb27-81" data-line-number="81"><span class="st">    &quot;</span></a>
<a class="sourceLine" id="cb27-82" data-line-number="82"></a>
<a class="sourceLine" id="cb27-83" data-line-number="83">    mRNA_near_synonyms &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">'transcript'</span>, <span class="st">'SO:0000673'</span>)</a>
<a class="sourceLine" id="cb27-84" data-line-number="84">    exon_near_synonyms &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">'SO:0000147'</span>, <span class="st">'coding_exon'</span>, <span class="st">'coding exon'</span>, <span class="st">'SO:0000195'</span>)</a>
<a class="sourceLine" id="cb27-85" data-line-number="85"></a>
<a class="sourceLine" id="cb27-86" data-line-number="86">    <span class="cf">if</span>(<span class="kw">any</span>(.<span class="op">$</span>type <span class="op">%in%</span><span class="st"> </span>mRNA_near_synonyms)){</a>
<a class="sourceLine" id="cb27-87" data-line-number="87">        .<span class="op">$</span>type &lt;-<span class="st"> </span><span class="kw">ifelse</span>(.<span class="op">$</span>type <span class="op">%in%</span><span class="st"> </span>mRNA_near_synonyms, <span class="st">'mRNA'</span>, .<span class="op">$</span>type)</a>
<a class="sourceLine" id="cb27-88" data-line-number="88">        <span class="kw">warning</span>(<span class="st">&quot;Substituting transcript types for mRNA types, this is probably OK&quot;</span>)</a>
<a class="sourceLine" id="cb27-89" data-line-number="89">    }</a>
<a class="sourceLine" id="cb27-90" data-line-number="90"></a>
<a class="sourceLine" id="cb27-91" data-line-number="91">    <span class="cf">if</span>(<span class="kw">any</span>(.<span class="op">$</span>type <span class="op">%in%</span><span class="st"> </span>exon_near_synonyms)){</a>
<a class="sourceLine" id="cb27-92" data-line-number="92">        .<span class="op">$</span>type &lt;-<span class="st"> </span><span class="kw">ifelse</span>(.<span class="op">$</span>type <span class="op">%in%</span><span class="st"> </span>exon_near_synonyms, <span class="st">'exon'</span>, .<span class="op">$</span>type)</a>
<a class="sourceLine" id="cb27-93" data-line-number="93">        <span class="kw">warning</span>(<span class="st">&quot;Substituting transcript types for exon types, this is probably OK&quot;</span>)</a>
<a class="sourceLine" id="cb27-94" data-line-number="94">    }</a>
<a class="sourceLine" id="cb27-95" data-line-number="95"></a>
<a class="sourceLine" id="cb27-96" data-line-number="96">  } <span class="op">%&gt;&gt;%</span><span class="st"> </span>{</a>
<a class="sourceLine" id="cb27-97" data-line-number="97"></a>
<a class="sourceLine" id="cb27-98" data-line-number="98">    <span class="st">&quot;</span></a>
<a class="sourceLine" id="cb27-99" data-line-number="99"><span class="st">    Notice here that I use the magrittr operator '%&gt;%' inside the rmonad</span></a>
<a class="sourceLine" id="cb27-100" data-line-number="100"><span class="st">    pipeline. When to pipe with rmonad and when to pipe with magrittr is a</span></a>
<a class="sourceLine" id="cb27-101" data-line-number="101"><span class="st">    matter of granularity. This chunk of code perhpas should form one</span></a>
<a class="sourceLine" id="cb27-102" data-line-number="102"><span class="st">    documentation unit. And perhaps I don't expect it to fail. If I break this</span></a>
<a class="sourceLine" id="cb27-103" data-line-number="103"><span class="st">    chunk into several, the failures are more localized, and I can access</span></a>
<a class="sourceLine" id="cb27-104" data-line-number="104"><span class="st">    intermediate values for debugging. On the other hand, putting every little</span></a>
<a class="sourceLine" id="cb27-105" data-line-number="105"><span class="st">    operation in a new chunk will clutter the graph and reports generated from</span></a>
<a class="sourceLine" id="cb27-106" data-line-number="106"><span class="st">    it.</span></a>
<a class="sourceLine" id="cb27-107" data-line-number="107"><span class="st">    &quot;</span></a>
<a class="sourceLine" id="cb27-108" data-line-number="108"></a>
<a class="sourceLine" id="cb27-109" data-line-number="109">    <span class="kw">data_frame</span>(</a>
<a class="sourceLine" id="cb27-110" data-line-number="110">      <span class="dt">attr  =</span> stringr<span class="op">::</span><span class="kw">str_split</span>(.<span class="op">$</span>attr, <span class="st">&quot;;&quot;</span>),</a>
<a class="sourceLine" id="cb27-111" data-line-number="111">      <span class="dt">order =</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(.)</a>
<a class="sourceLine" id="cb27-112" data-line-number="112">    ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb27-113" data-line-number="113"><span class="st">      </span>dplyr<span class="op">::</span><span class="kw">mutate</span>(<span class="dt">ntags =</span> <span class="kw">sapply</span>(attr, length)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb27-114" data-line-number="114"><span class="st">      </span>tidyr<span class="op">::</span><span class="kw">unnest</span>(attr) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb27-115" data-line-number="115"><span class="st">      </span>dplyr<span class="op">::</span><span class="kw">mutate</span>(<span class="dt">attr =</span> <span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">'='</span>, attr), attr, <span class="kw">paste</span>(<span class="st">&quot;.U&quot;</span>, attr, <span class="dt">sep=</span><span class="st">&quot;=&quot;</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb27-116" data-line-number="116"><span class="st">      </span>tidyr<span class="op">::</span><span class="kw">separate_</span>(</a>
<a class="sourceLine" id="cb27-117" data-line-number="117">        <span class="dt">col   =</span> <span class="st">&quot;attr&quot;</span>,</a>
<a class="sourceLine" id="cb27-118" data-line-number="118">        <span class="dt">into  =</span> <span class="kw">c</span>(<span class="st">&quot;tag&quot;</span>, <span class="st">&quot;value&quot;</span>),</a>
<a class="sourceLine" id="cb27-119" data-line-number="119">        <span class="dt">sep   =</span> <span class="st">&quot;=&quot;</span>,</a>
<a class="sourceLine" id="cb27-120" data-line-number="120">        <span class="dt">extra =</span> <span class="st">&quot;merge&quot;</span></a>
<a class="sourceLine" id="cb27-121" data-line-number="121">      )</a>
<a class="sourceLine" id="cb27-122" data-line-number="122"></a>
<a class="sourceLine" id="cb27-123" data-line-number="123">   } <span class="op">%v&gt;%</span><span class="st"> </span><span class="kw">funnel</span>(<span class="dt">raw_gff=</span>raw_gff, <span class="dt">tags=</span>tags) <span class="op">%*&gt;%</span><span class="st"> </span>{</a>
<a class="sourceLine" id="cb27-124" data-line-number="124"></a>
<a class="sourceLine" id="cb27-125" data-line-number="125">    <span class="st">&quot;</span></a>
<a class="sourceLine" id="cb27-126" data-line-number="126"><span class="st">    The %v&gt;% operator stores the input value. We could replace every %&gt;&gt;%</span></a>
<a class="sourceLine" id="cb27-127" data-line-number="127"><span class="st">    operator with %v&gt;%. This would let us inspect every step of an analysis at</span></a>
<a class="sourceLine" id="cb27-128" data-line-number="128"><span class="st">    the cost of high memory usage. For brevity, I won't break this following</span></a>
<a class="sourceLine" id="cb27-129" data-line-number="129"><span class="st">    block down any further.</span></a>
<a class="sourceLine" id="cb27-130" data-line-number="130"></a>
<a class="sourceLine" id="cb27-131" data-line-number="131"><span class="st">    The `funnel` function packages a list in a monad, merging their histories</span></a>
<a class="sourceLine" id="cb27-132" data-line-number="132"><span class="st">    and propagating error. That is, if `gff` or `tags` failed upstream, this</span></a>
<a class="sourceLine" id="cb27-133" data-line-number="133"><span class="st">    function will not be run. `%*&gt;%` takes a list on the left and feeds it into</span></a>
<a class="sourceLine" id="cb27-134" data-line-number="134"><span class="st">    the function on the right as an argument list. Here `funnel` and `%*&gt;%` are</span></a>
<a class="sourceLine" id="cb27-135" data-line-number="135"><span class="st">    used together to merge a pipeline (gff) and inject a parameter (tags).</span></a>
<a class="sourceLine" id="cb27-136" data-line-number="136"></a>
<a class="sourceLine" id="cb27-137" data-line-number="137"><span class="st">    We could not have written</span></a>
<a class="sourceLine" id="cb27-138" data-line-number="138"></a>
<a class="sourceLine" id="cb27-139" data-line-number="139"><span class="st">      %v&gt;% function(gff=gff, tags=tags)</span></a>
<a class="sourceLine" id="cb27-140" data-line-number="140"></a>
<a class="sourceLine" id="cb27-141" data-line-number="141"><span class="st">    because this would have brough the monad wrapped gff into scope, not the</span></a>
<a class="sourceLine" id="cb27-142" data-line-number="142"><span class="st">    value itself.</span></a>
<a class="sourceLine" id="cb27-143" data-line-number="143"><span class="st">    &quot;</span></a>
<a class="sourceLine" id="cb27-144" data-line-number="144"></a>
<a class="sourceLine" id="cb27-145" data-line-number="145">      dplyr<span class="op">::</span><span class="kw">filter</span>(., tag <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(tags, <span class="st">&quot;.U&quot;</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb27-146" data-line-number="146"><span class="st">      </span>{</a>
<a class="sourceLine" id="cb27-147" data-line-number="147">        <span class="cf">if</span>(<span class="kw">nrow</span>(.) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>){</a>
<a class="sourceLine" id="cb27-148" data-line-number="148">          tidyr<span class="op">::</span><span class="kw">spread</span>(., <span class="dt">key=</span><span class="st">&quot;tag&quot;</span>, <span class="dt">value=</span><span class="st">&quot;value&quot;</span>)</a>
<a class="sourceLine" id="cb27-149" data-line-number="149">        } <span class="cf">else</span> {</a>
<a class="sourceLine" id="cb27-150" data-line-number="150">          .<span class="op">$</span>tag   =<span class="st"> </span><span class="ot">NULL</span></a>
<a class="sourceLine" id="cb27-151" data-line-number="151">          .<span class="op">$</span>value =<span class="st"> </span><span class="ot">NULL</span></a>
<a class="sourceLine" id="cb27-152" data-line-number="152">          .</a>
<a class="sourceLine" id="cb27-153" data-line-number="153">        }</a>
<a class="sourceLine" id="cb27-154" data-line-number="154">      } <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb27-155" data-line-number="155"><span class="st">      </span>{</a>
<a class="sourceLine" id="cb27-156" data-line-number="156">        <span class="cf">if</span>(<span class="st">&quot;Parent&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">names</span>(.)){</a>
<a class="sourceLine" id="cb27-157" data-line-number="157">          .<span class="op">$</span>Parent &lt;-<span class="st"> </span><span class="kw">ifelse</span>(.<span class="op">$</span>Parent <span class="op">==</span><span class="st"> &quot;-&quot;</span>, <span class="ot">NA</span>, .<span class="op">$</span>Parent)</a>
<a class="sourceLine" id="cb27-158" data-line-number="158">        }</a>
<a class="sourceLine" id="cb27-159" data-line-number="159">        .</a>
<a class="sourceLine" id="cb27-160" data-line-number="160">      } <span class="op">%&gt;%</span><span class="st"> </span>{</a>
<a class="sourceLine" id="cb27-161" data-line-number="161">        <span class="cf">for</span>(tag <span class="cf">in</span> <span class="kw">c</span>(tags, <span class="st">&quot;.U&quot;</span>)){</a>
<a class="sourceLine" id="cb27-162" data-line-number="162">          <span class="cf">if</span>(<span class="op">!</span><span class="st"> </span>tag <span class="op">%in%</span><span class="st"> </span><span class="kw">names</span>(.))</a>
<a class="sourceLine" id="cb27-163" data-line-number="163">            .[[tag]] =<span class="st"> </span><span class="ot">NA_character_</span></a>
<a class="sourceLine" id="cb27-164" data-line-number="164">        }</a>
<a class="sourceLine" id="cb27-165" data-line-number="165">        .</a>
<a class="sourceLine" id="cb27-166" data-line-number="166">      } <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb27-167" data-line-number="167"><span class="st">      </span>{</a>
<a class="sourceLine" id="cb27-168" data-line-number="168">        <span class="cf">if</span>(<span class="st">&quot;ID&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">names</span>(.))</a>
<a class="sourceLine" id="cb27-169" data-line-number="169">          .<span class="op">$</span>ID &lt;-<span class="st"> </span><span class="kw">ifelse</span>(<span class="kw">is.na</span>(.<span class="op">$</span>ID) <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(.<span class="op">$</span>.U) <span class="op">&amp;</span><span class="st"> </span>.<span class="op">$</span>ntags <span class="op">==</span><span class="st"> </span><span class="dv">1</span>, .<span class="op">$</span>.U, .<span class="op">$</span>ID)</a>
<a class="sourceLine" id="cb27-170" data-line-number="170">        .</a>
<a class="sourceLine" id="cb27-171" data-line-number="171">      } <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb27-172" data-line-number="172"><span class="st">      </span><span class="kw">merge</span>(<span class="kw">data_frame</span>(<span class="dt">order=</span><span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(raw_gff)), <span class="dt">all=</span><span class="ot">TRUE</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb27-173" data-line-number="173"><span class="st">      </span>dplyr<span class="op">::</span><span class="kw">arrange</span>(order) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb27-174" data-line-number="174"><span class="st">      </span>{ <span class="kw">cbind</span>(raw_gff, .) } <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb27-175" data-line-number="175"><span class="st">      </span>dplyr<span class="op">::</span><span class="kw">select</span>(<span class="op">-</span>.U, <span class="op">-</span>order, <span class="op">-</span>ntags, <span class="op">-</span>attr)</a>
<a class="sourceLine" id="cb27-176" data-line-number="176"></a>
<a class="sourceLine" id="cb27-177" data-line-number="177">  } <span class="op">%&gt;_%</span><span class="st"> </span>{</a>
<a class="sourceLine" id="cb27-178" data-line-number="178"></a>
<a class="sourceLine" id="cb27-179" data-line-number="179">    <span class="st">&quot;</span></a>
<a class="sourceLine" id="cb27-180" data-line-number="180"><span class="st">    And make the last few assertions.</span></a>
<a class="sourceLine" id="cb27-181" data-line-number="181"><span class="st">    &quot;</span></a>
<a class="sourceLine" id="cb27-182" data-line-number="182"></a>
<a class="sourceLine" id="cb27-183" data-line-number="183">    <span class="cf">if</span>(<span class="kw">all</span>(<span class="kw">c</span>(<span class="st">&quot;ID&quot;</span>, <span class="st">&quot;Parent&quot;</span>) <span class="op">%in%</span><span class="st"> </span><span class="kw">names</span>(.))){</a>
<a class="sourceLine" id="cb27-184" data-line-number="184">      parents &lt;-<span class="st"> </span><span class="kw">subset</span>(., type <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;CDS&quot;</span>, <span class="st">&quot;exon&quot;</span>))<span class="op">$</span>Parent</a>
<a class="sourceLine" id="cb27-185" data-line-number="185">      parent_types &lt;-<span class="st"> </span><span class="kw">subset</span>(., ID <span class="op">%in%</span><span class="st"> </span>parents)<span class="op">$</span>type</a>
<a class="sourceLine" id="cb27-186" data-line-number="186"></a>
<a class="sourceLine" id="cb27-187" data-line-number="187">      <span class="cf">if</span>(<span class="kw">any</span>(parent_types <span class="op">==</span><span class="st"> &quot;gene&quot;</span>))</a>
<a class="sourceLine" id="cb27-188" data-line-number="188">        <span class="kw">warning</span>(<span class="st">&quot;Found CDS or exon directly inheriting from a gene, this may be fine.&quot;</span>)</a>
<a class="sourceLine" id="cb27-189" data-line-number="189"></a>
<a class="sourceLine" id="cb27-190" data-line-number="190">      <span class="cf">if</span>(<span class="op">!</span><span class="st"> </span><span class="kw">all</span>(parent_types <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;gene&quot;</span>, <span class="st">&quot;mRNA&quot;</span>)))</a>
<a class="sourceLine" id="cb27-191" data-line-number="191">        <span class="kw">stop</span>(<span class="st">&quot;Found CDS or exon with illegal parent&quot;</span>)</a>
<a class="sourceLine" id="cb27-192" data-line-number="192"></a>
<a class="sourceLine" id="cb27-193" data-line-number="193">      <span class="cf">if</span>( <span class="kw">any</span>(<span class="kw">is.na</span>(parents)) )</a>
<a class="sourceLine" id="cb27-194" data-line-number="194">        <span class="kw">stop</span>(<span class="st">&quot;Found CDS or exon with no parent&quot;</span>)</a>
<a class="sourceLine" id="cb27-195" data-line-number="195"></a>
<a class="sourceLine" id="cb27-196" data-line-number="196">      <span class="cf">if</span>(<span class="op">!</span><span class="st"> </span><span class="kw">any</span>(<span class="kw">duplicated</span>(.<span class="op">$</span>ID, <span class="dt">incomparables=</span><span class="ot">NA</span>)))</a>
<a class="sourceLine" id="cb27-197" data-line-number="197">        <span class="kw">warning</span>(<span class="st">&quot;IDs are not unique, this is probably bad&quot;</span>)</a>
<a class="sourceLine" id="cb27-198" data-line-number="198">    }</a>
<a class="sourceLine" id="cb27-199" data-line-number="199"></a>
<a class="sourceLine" id="cb27-200" data-line-number="200">  } <span class="op">%&gt;_%</span><span class="st"> </span>{</a>
<a class="sourceLine" id="cb27-201" data-line-number="201"></a>
<a class="sourceLine" id="cb27-202" data-line-number="202">    <span class="st">&quot;</span></a>
<a class="sourceLine" id="cb27-203" data-line-number="203"><span class="st">    I could post some closing comments here. The %&gt;_% operator can be chained</span></a>
<a class="sourceLine" id="cb27-204" data-line-number="204"><span class="st">    and the output does not affect the output of the main chain. The NULL is</span></a>
<a class="sourceLine" id="cb27-205" data-line-number="205"><span class="st">    required to distinguish this block from an anonymous function that returns</span></a>
<a class="sourceLine" id="cb27-206" data-line-number="206"><span class="st">    a string.</span></a>
<a class="sourceLine" id="cb27-207" data-line-number="207"><span class="st">    &quot;</span></a>
<a class="sourceLine" id="cb27-208" data-line-number="208"></a>
<a class="sourceLine" id="cb27-209" data-line-number="209">    <span class="ot">NULL</span></a>
<a class="sourceLine" id="cb27-210" data-line-number="210">  }</a>
<a class="sourceLine" id="cb27-211" data-line-number="211">  <span class="co"># End Rmonad chain</span></a>
<a class="sourceLine" id="cb27-212" data-line-number="212"></a>
<a class="sourceLine" id="cb27-213" data-line-number="213">}</a></code></pre>
<p>That is the whole GFF program in an rmonad framework</p>
<pre class="sourceCode r" id="cb28"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1">result &lt;-<span class="st"> </span><span class="kw">read_gff</span>(<span class="dt">file=</span>gff<span class="op">$</span>good, <span class="dt">tags=</span><span class="kw">c</span>(<span class="st">&quot;ID&quot;</span>, <span class="st">&quot;Parent&quot;</span>))</a></code></pre>
<p><code>esc</code> will extract the final result, and raise all errors, warnings and messages that where extracted</p>
<pre class="sourceCode r" id="cb29"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1"><span class="kw">esc</span>(result)</a></code></pre>
<pre><code>## Warning: in 'function (.) 
## {
##     if (all(c(&quot;ID&quot;, &quot;Parent&quot;) %in% names(.))) {
##         parents &lt;- subset(., type %in% c(&quot;CDS&quot;, &quot;exon&quot;))$Parent
##         parent_types &lt;- subset(., ID %in% parents)$type
##         if (any(parent_types == &quot;gene&quot;)) 
##             warning(&quot;Found CDS or exon directly inheriting from a gene, this may be fine.&quot;)
##         if (!all(parent_types %in% c(&quot;gene&quot;, &quot;mRNA&quot;))) 
##             stop(&quot;Found CDS or exon with illegal parent&quot;)
##         if (any(is.na(parents))) 
##             stop(&quot;Found CDS or exon with no parent&quot;)
##         if (!any(duplicated(.$ID, incomparables = NA))) 
##             warning(&quot;IDs are not unique, this is probably bad&quot;)
##     }
## }': Found CDS or exon directly inheriting from a gene, this may be fine.</code></pre>
<pre><code>##    seqid source type start  stop score strand phase    ID Parent
## 1      I   &lt;NA&gt; gene 11565 11951    NA      -    NA gene0   &lt;NA&gt;
## 2      I   &lt;NA&gt; mRNA 11565 11951    NA      -    NA mrna0  gene0
## 3      I   &lt;NA&gt; exon 11565 11951    NA      -    NA  &lt;NA&gt;  mrna0
## 4      I   &lt;NA&gt;  CDS 11565 11951    NA      -     0  &lt;NA&gt;  mrna0
## 5      I   &lt;NA&gt; gene 61931 83591    NA      +    NA gene1   &lt;NA&gt;
## 6      I   &lt;NA&gt; mRNA 61931 83591    NA      +    NA  rna1  gene1
## 7      I   &lt;NA&gt; exon 61931 62344    NA      +    NA  &lt;NA&gt;   rna1
## 8      I   &lt;NA&gt; exon 81616 82209    NA      +    NA  &lt;NA&gt;   rna1
## 9      I   &lt;NA&gt; exon 82211 83591    NA      +    NA  &lt;NA&gt;   rna1
## 10     I   &lt;NA&gt;  CDS 61931 62344    NA      +     0  &lt;NA&gt;   rna1
## 11     I   &lt;NA&gt;  CDS 81616 82209    NA      +     0  &lt;NA&gt;   rna1
## 12     I   &lt;NA&gt;  CDS 82211 82681    NA      +     0  &lt;NA&gt;   rna1
## 13     I   &lt;NA&gt; mRNA 61931 83591    NA      +    NA  rna2  gene1
## 14     I   &lt;NA&gt; exon 61931 62344    NA      +    NA  &lt;NA&gt;   rna2
## 15     I   &lt;NA&gt; exon 81616 82209    NA      +    NA  &lt;NA&gt;   rna2
## 16     I   &lt;NA&gt; exon 82211 83591    NA      +    NA  &lt;NA&gt;   rna2
## 17     I   &lt;NA&gt;  CDS 61931 62344    NA      +     0  &lt;NA&gt;   rna2
## 18     I   &lt;NA&gt;  CDS 82211 82681    NA      +     0  &lt;NA&gt;   rna2
## 19     b   &lt;NA&gt; gene  7235  9016    NA      -    NA gene2   &lt;NA&gt;
## 20     b   &lt;NA&gt;  CDS  7235  9016    NA      -     0  &lt;NA&gt;  gene2
## 21     I   &lt;NA&gt; gene 11565 11951    NA      -    NA gene3   &lt;NA&gt;
## 22     I   &lt;NA&gt; gene 11565 11951    NA      -    NA gene3   &lt;NA&gt;
## 23     I   &lt;NA&gt; mRNA 11565 11951    NA      -    NA mrna3  gene3
## 24     I   &lt;NA&gt; exon 11565 11951    NA      -    NA  &lt;NA&gt;  mrna3
## 25     I   &lt;NA&gt;  CDS 11565 11951    NA      -     0  &lt;NA&gt;  mrna3</code></pre>
<p>Now we see why we might want a little more granularity in our pipeline. To summarize the results we can use the <code>mtabulate</code> functions.</p>
<pre class="sourceCode r" id="cb32"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1"><span class="kw">mtabulate</span>(result)</a></code></pre>
<pre><code>##   id   OK cached time space is_nested ndependents nnotes nwarnings error
## 1  1 TRUE  FALSE    0   152         0           1      0         0     0
## 2  2 TRUE  FALSE    0  9432         0           1      0         0     0
## 3  3 TRUE  FALSE    0  2416         0           1      0         0     0
## 4  4 TRUE  FALSE    0 12400         0           0      0         0     0
## 5  5 TRUE  FALSE    0 12400         1           1      0         0     0
## 6  6 TRUE  FALSE    0  4640         0           1      0         0     0
## 7  7 TRUE  FALSE    0  4640         0           1      0         1     0
## 8  8 TRUE   TRUE    0  4640         0           0      0         0     0
##   doc
## 1   0
## 2   1
## 3   0
## 4   0
## 5   0
## 6   1
## 7   1
## 8   1</code></pre>
<p>We can also get a summary of issues</p>
<pre class="sourceCode r" id="cb34"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1"><span class="kw">missues</span>(result)</a></code></pre>
<pre><code>##   id    type
## 1  7 warning
##                                                                  issue
## 1 Found CDS or exon directly inheriting from a gene, this may be fine.</code></pre>
<p>The <code>id</code> column corresponds to a row number in the <code>mtabulate</code> result.</p>
<p>To extract particular values, we can use the <code>get_*</code> family of vectorized getters</p>
<pre class="sourceCode r" id="cb36"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1"><span class="co"># get a list of every stored value, report uncached values as NULL</span></a>
<a class="sourceLine" id="cb36-2" data-line-number="2"><span class="kw">get_value</span>(result, <span class="dt">warn=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb36-3" data-line-number="3"><span class="co"># get a list of every docstring</span></a>
<a class="sourceLine" id="cb36-4" data-line-number="4"><span class="kw">get_doc</span>(result)</a></code></pre>
<pre class="sourceCode r" id="cb37"><code class="sourceCode r"><a class="sourceLine" id="cb37-1" data-line-number="1"><span class="kw">plot</span>(result)</a></code></pre>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAkAAAAJACAIAAADq+E5hAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAgAElEQVR4nO3dd1xW5eP/8etmC8pImaI4ksQUnJk4cCFOUBsiKWlm2tDSr5/KypRsYGWGpubKVWqhKC4cOXGnhkquxM0GWSLewM39+8N+fvooNyLe67p5PR/9Aedc5/C2Tx/fXOe+zjkKtVotAACQjZmhAwAAUBUUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShaGDgAAlaJUKnft2hUbG3vmzJmbN29mZ2c7Ojp6eHh4e3sPGDCgT58+Dg4Ohs4IvVKo1WpDZwCAihQVFc2ePfvbb799qtlTfsF+Dds3dPRwrFmn5p3cO/lp+TdP30zYmHAp/tLr4a9PmTKlTp06hs4LPaHAABi1AwcOhIWFubR3CZke4vK0i6Zht7Nub/1y6+nVp7/77rvhw4frMyEMhQIDYLx++umnCR9NGLF0RLPAZpUZn3oudf7g+a8OejUyMtLMjM/4TRwFBsBIrVq16t1P3h23ZZxrE9fKH1V4q3D+C/MHPj/wm2++0V02GAN+QwFgjI4dOzZ2wti3Nrz1WO0lhLB7yu7NdW+u2LBixYoVOsoGI8EMDIDRUalULVu27DilY6vBrap2htRzqXO7zz179qyzs7N2s8F4MAMDYHSWL1+udFJWub2EEO4+7r5DfT///HMtpoKxYQYGwOj4+PgMWDCgsX/jcvfezlJHT1ImxpVa2SoC/8+y+ztWGobd/tTn0+QryY6OjroMC4NhBgbAuJw7dy6jKKNRh0aaBmz4RNn1TasZ12t2H2f52wRlcmJZucNq1qnZpEuTuLg4nSWFgVFgAIzL5s2b/fr7KRSKcvfmpah7TbJq2N7Mwlr0nGBlYS3u5Gi8jOQX7Ldp0yadJYWBUWAAjMv58+frtaynaa+Dh8Ll6X/+4spMKvNqY97oeXNNgz19PS9cuKD9iDAOFBgA45KWlmbvav/IYel/ly1+pShkupW5pcYxju6OKSkp2gwHY0KBATAuOTk5tk62FY+5ely19Utlbop6VmDRn+tLNQ2zdbLNycnRdkAYCwoMgHFxdnYuyCioeEyDtuYjl9b45LidRzOz3XOKNQ0ryChwdX28+6AhEQoMgHFxd3fPS8urzEh7V0XHUZa3szQu4shLy3N3d9deNBgXCgyAcfHz87t85HIlB9u7KLzaalzEkXQ4yc/PT0u5YHQoMADGJSQkJDEusay0/Lu77uSoT0SXFt9RCyHKVCJhY2nwVGtNpzoVe2rgwIG6CgpD40kcAIyOv7+/7wTfVoPKeZRUfpr6h+Ci3NSygDFWjnUVzYMsHDzKv2Ms41LG7M6zb9y4YWNjo+O8MAwKDIDR2bFjR/i48KmnpppZVP0q0cLQhUOeG/L+++9rMRiMCgUGwBgFBgba97DvNalX1Q4/u+Pshjc3nD9/vkaNGtoNBuPBZ2AAjNHy5cv/mPfHmS1nqnBs1tWs5a8vX7FiBe1l2igwAMbIw8NjzZo1K0av+Gv7X491YMrZlO97ff/d598FBAToKBuMBJcQARivQ4cOvfTSS8+Pe77nez0r83nYyXUnV49f/eOsH8PCwvQQD4ZFgQEwajdu3HjttdcSbySGTA/xC/YzMy+/xpIOJcVOibXOsV6yZEm7du30HBIGQYEBkMD27ds/++yz0xdPN+/TvHGHxg7uDvau9rezbudn5F8/ef30ltPONs4ffvhheHi4mRmfjFQXFBgAaSQnJ2/atOnUqVPJycnp6el16tRxc3Pz8fHp379/06ZNDZ0O+kaBAZCVSqU6fPhwp06dDB0EhkGBAZBVVlaWj49PZmamoYPAMLhYDEBWderUKSoqun37tqGDwDAoMAASq1ev3o0bNwydAoZBgQGQmJeX17Vr1wydAoZBgQGQGAVWnVFgACRWv359CqzaosAASIwZWHVGgQGQGAVWnVFgACRGgVVn3MgMQGIqlcrOzq6goMDS0tLQWaBvzMAASMzc3NzNze3mzZuGDgIDoMAAyI2riNUWBQZAbhRYtUWBAZAbBVZtUWAA5Obl5XX9+nVDp4ABWBg6AAA8ES8vr19//dXQKbQjKytr586d165dS01NVSqVbm5ubm5u/v7+vr6+ho5mjFhGD0BuFy9e7N+//8WLFw0dpOrUavXatWvnzp37x6k/vLt6uzZxdfBwsLSxzE/Lz03OPb/7fC1Ra+jQof/3f//n5ORk6LBGhAIDILe7d+86OTnduXNHoVAYOktVHDt2bNy4cemq9H4f9/MJ9LG0KeeGtpS/UvbM3XMh9sJHH300fvx4Sf+kWkeBAZCem5tbQkKCm5uboYM8thUrVox7f9xL377UdkjbR9ZS2oW0lW+sbOXRatmyZXZ2dvpJaMxYxAFAepIuRJw5c+Z/pv9n4q6J7ULbVWZS5faM24QdE1LsUrp3737nzh09JDRyFBgA6cn4UpW4uLgvvv9iws4Jbs88xsTRwtoifHG4VXOr4cOHc/2MAgMgPelmYKmpqa+++uqY6DFOnlVZlBE6OzQhJWH+/PlaDyYXCgyA9KQrsIiIiNYjWjdo26Bqh1tYW4QvCp8+fXp+fr5Wc0mGAgMgPbkK7MqVK6vWrwr6T5CmAaoSEb+oZNHQuxWcxK2pW8NeDWfPnq2DgNKgwABIT64Ci46ObvNiG1sn23L33s0XJ9eV7p1XfOuGquLzBIwNWLNmjQ4CSoMCAyA9uQosNjbWL9hP014be9Eu1KJRB/NHnserrVdKXorUd3A/IQoMgPQcHR0VCkVubq6hgzxaWVnZiRMnmnRqUvGwytyprFAomnRu8scff2gnmYQoMACmQJaV9JmZmdaO1hbW2nkOrWNdx+TkZK2cSkYUGABTIMsz6dPT02u51NLW2RzcHNLS0rR1NulQYABMgSwfg9nZ2RUXFmvrbMpCZXV+phQFBsAUyFJg7u7uuala+6wuLyWvbt262jqbdCgwAKZAlgKztbV9yv6pW9duaeVsaRfSGjZsqJVTyYgCA2AKZFnEIYTo27fvqc2nKh6jvKNWlz1iJWLhrcL0hPSAgADtRZMMBQbAFMgyAxNCDBw48OS6k5r2lirF0V9Kzu5Q3TytOvpLSWG2xif2ntp0qlu3bjY2NrqJKQHeBwbAFKjValtb21u3btWoUcPQWR6htLTU19e316xePj18qn6S4tII34g1C9d0795di9nkwgwMgClQKBSenp43btwwdJBHs7Cw+Oyzz9Z/tL6stKzKJ9k3f197n/bVub0EBQbAZEh0FfGFF15o4d7it//7rWqHJx1O2v/N/u+++067qaRDgQEwERIVWEFBwfLly9P3pP/+/e+Pe2za+bRFQxctX768SZNHPI/K5FFgAEyELAV2+PBhHx+fCxcubN269dyyc6vHra78tcTEuMSZPWb+MOOHoCCNb2OpPigwACZCigKLjo4OCQn58ccf/f39GzRocOjQIYcUhy+e+yJxW2LFB+bcyFk+avmGtzbsiN3xyiuv6CetkdPOAyUBwOCM/HGIarX666+/njt37vbt21u1anVvo729fWxs7MaNGyd/MHnzZ5tbD27t29/XubGzueU/r1MpyCi4sOdCwsaEq7uvvvXWW5POTqpVS2uPUpQdy+gBmIgrV6507979ypUrhg5SDqVS+frrr1+6dGn9+vVubm4PD1CpVHv27NmwYcO2bdtu3Lhh85SNhZVFfnq+k71Thw4dQkJCQkJCateurf/kxowCA2AiSkpKatWqdfv2bQsL47q2lJWVNXjwYDc3t+XLl1fmNjW1Wp2enl5cXOzq6mptba2HhJLiMzAAJsLS0tLZ2Tk1NdXQQf5HYmLic8895+/vv2bNmkreZK1QKNzc3OrXr097VYwCA2A6jG0dx/bt23v06PHZZ59FRkaamfH3rZbxLxSA6TCqAlu4cOGIESPWrl07bNgwQ2cxTcZ1pRgAnoSRPJO+tLR0woQJe/fuPXz4cIMGDQwdx2RRYABMh5eX159//mnYDAUFBaGhoSqV6sCBAw4ODoYNY9q4hAjAdDxwCbG0tPT69ev6XNZx+fLl9u3be3p6bt68mfbSNQoMgPTu3r174cKFHTt2/PHHH8ePHw8PD+/SpUuDBg1q1qzp5eX18ccf6yfGoUOHOnXq9NZbby1YsMDYlvKbJO4DAyCxIUOG7Nu3Lz09vYIx+/bt69Kli66TrFmzZvz48cuWLevbt6+ufxbu4XcEABLr0KHDb79V9FKSZs2a6bq91Gp1RETEzz//vG/fPh+fqr+jEo+LS4gAJPbuu+8OGDCgggEjR47UaYC7d++GhYXt3Lnz0KFDtJeeUWAAJKZQKBYvXuzu7q5pQGhoqO5+ekpKSufOnYUQu3btcnFx0d0PQrkoMAByc3Fx+fbbb8vd1aFDB09PTx393FOnTvn7+/fr12/16tU2NjY6+imoAAUGQHphYWGBgYEPb+/Xr5+OfuL69et79OgRGRk5bdo0Hf0IPBIFBsAUzJo16+GNvXr1qvIJV65c+c0335S7Kyoqavz48du2bdPp9Uk8EsvoAZiI0aNHL168+P63tWrVys7OtrS0rNrZWrVqlZCQsGrVqqFDh97fWFxcPGbMmD///HPjxo3169d/0sR4MszAAJiIjz766N/ftm7dusrtdfLkyYSEBCFEWFjYkSNH7m28detWUFBQXl7ewYMHaS9jQIEBMBENGzYMDw+//+2zzz5b5VMtWLDg/tcvvPDCtWvXLl265O/v7+vru3btWjs7uycKCi2hwACYjrfeeuv+197e3lU7SVZWVnR09P1vU1JSunXr1rFjxw8//DAqKorXehkP/pcAYDrat2/fqlWre18//fTTVTvJ3Llzc3Jy/r3lypUrDRs2HD58+JPmg1ZRYABMyssvv3zvi6pd6MvLy5s7d+7D248ePTpp0qQnSgZto8AAmJR7j8YQQpSVlVXh8MjIyMzMzHJ3ff/99z/++GPVk0HbWEYPwKQolUpHR8e7d+/+/vvvPXr0eKxjk5KSHnnhcefOnT179nyCgNAaZmAATIq1tXXTpk3F48/A1Gr1uHHjHjlsyJAhKpWqiuGgVbxOBYCp6dy5c0JCwuMW2OLFi+Pi4srdVbNmzdatW7dv375NmzatW7c2NzfXRkw8KQoMgKkJCgqaM2dOcXFx5Q9JSkp6+MXN9erVCwkJCQ4O7tSpU40aNbSaEVpAgQEwKSqV6t4DOFauXHn8+HF3d3cPD4927dpV8MqV9PT0Pn36/HvtxrBhw0aMGNGtWzfu+jJmLOIAYCJOnDgxe/bsrVu32tazdW/mbu9ib2Nvk5+Wn5Ock3Q4qfnTzYcOHfrGG288MJfKycnp3LnzX3/9JYR46qmnRo0a9d5773l4eBjoD4HHQIEBkN7169c/+OCD7Qe2B04MbD24tWNdxwcGlJWWXdx/cf/C/Vl/ZE2fPv3+E6fu3LnTuXPnkydPCiH69eu3cOFCqksiFBgAucXHxw8ZMqTtmLaBEwKtbK0qHnzl2JVVb63q2brn/Pnz1Wp1hw4dEhISHBwcFi5ceP8OaMiCAgMgsfXr1494a8TIZSN9evhU8hBloXL5qOV2GXa5ublnzpzx9/ePjo5m4iUjCgyArP7888+A3gHjt46v51fvsQ5Uq9U/hf90/Nfjc+bMefvtt3UUD7rGKkQAUrpz587gwYPDfgh73PYSQigUivCF4ZmXMrmjS2qsEAUgpe+++86lg0urQa2qdrhlDcsRS0dERETk5+drNxj0hgIDIJ/c3NzZs2cHRwRXPOzm6bIPG9zWtNetqVujoEazZs3SdjroCQUGQD6bNm3y7OxZp2GdCsaUFIlf3ioqVVZ0nu7ju//8889aDgd9ocAAyGfDhg0tg1tWPGbrV8p2Qx6xqt7T1zO3NPfs2bPaiwb9ocAAyOfgwYNNuzetYMDpTar6rcyd6ioeeaqmPZrGx8drLxr0hwIDIJmSkpLs3Gx7V3tNA/JS1ZePqFoNqtQqa6d6TikpKdpLB/2hwABIJj09vZZzLYVZ+bMrtVps/VLZ5yPLSp7Nwd2BApMUBQZAMlZWVqXFpZr27vmhxH+EpbXdoy8e3qMqVllZPeKjMhgnbmQGIBlnZ+fivOLS4lILq3L+Bju8vOS3if/zxuSxlgWBE61emGFd7tlyk3Ob122uk6DQMQoMgGQUCoWHh0fWlSy3Z9we3vvBoRpC/c/069SmklXvKGdcq2mm+YEbGUkZ9YPr6ygqdIpLiADkExQUdGbrmXJ3WVgpLKzFP/9YKRQKYWEtzDT8rq4qUZ3fdb5nz546zAqdocAAyCckJCRhfcKTn+f87vN+Tf3c3MqZycH4UWAA5NOjRw+zLLPzu89XPMwv2OLb1JoVDNj65daxY8dqNRr0hwIDIB9LS8svv/wyZnJMmaqsyif5c/2fte7UeuWVV7QYDPpEgQGQ0gsvvNC0TtP1k9dX7fCsK1lr3l0zZ84cMzP+GpQV/8sBkJJCoVizZs3VzVfjFz/2g6BuZ92eN2jeV1O+6tSpky6yQT9YRg9AQqWl4t13nXx9N2/e3Ldv3+wr2SHTQzQ9m+MByYnJ8wfPz7qSdfr0abVarVBU9pZnGBtmYABkc/u2GDhQXLokhg719vY+evTo3WN3Z/aYeeXYlYqPUxYqt3yx5fte38/5ck5wcPCPP/4YFBRUVFSkn9TQOoVarTZ0BgCotJQUMWCAaN1azJ8vLP65hqRSqVasWDF16lTHFo6tX2jdom+LmnX+u/hQXaa+cuzKqY2njvxypH+3/p9//nmDBg2USmXz5s0vXbrk6+u7bds2d3d3A/15UHUUGAB5nDkjBgwQI0aIadMe3llUVBQTExMbG7t9+3ZhKxxcHaxrWeel5uWm5DbzbhYcHDxkyJBnn332/vjr16/7+fnl5ua6u7tv3Lixbdu2+vuDQBsoMACS2LlTDBsmvv9eDB36yLHp6elpaWn5+fkeHh4eHh41atQod9jmzZsHDBgghKhdu/bnn3/+xhtvsChRIhQYABn89JP46CMRHS06d9buiSdPnhwZGXnv6zZt2sydO7d9+/ba/RHQEQoMgHFTq0VEhPj5Z7Fli3jmGa2fvrS0tH379idPnry/ZfDgwVOnTvX19dX6z4J2UWAAjJhSKV57TVy+LDZuFM7OOvohu3btevh5vkFBQePHj+/duzcXFY0WBQbAWN26JQYNEs7OYuVKoeFDLG3p3bv39u3bH95er1694cOHv/zyy35+fjoNgCqgwAAYpcuXRd++IihIzJoldD8HOnz4sL+/fwUDGjdu3K9fvy+//NLOzk7XYVBJTI0BGJ8jR0SnTmL8eBEVpYf2EkJ06NChQ4cOFQxISko6cOCAHpKg8igwAEYmJkYEB4tFi8Rbb+nzx44bN66CvR06dNi1axfTL6PCJUQAxiQqSnz7rYiNFa1b6/kn371719PTMzs7++Fd3bp127hxY82aFb1aDPrHDAyAcVCpxDvviCVLxMGD+m8vIYSNjc2gQYMe3t67d+8tW7bQXkaIAgNgBAoLxcCB4uJFER8v6tc3VIoXX3zxgS116tRp2LChpgd5wLAoMACGlpoqAgKEi4vYskU4OBgwSLdu3Rz+FSA0NPTChQvHjx+fOHGiAVNBEwoMgEElJooOHUT//mLJEmFpadgsVlZWAQEB974ePXr0L7/88tRTT+3YsWPv3r2ff/65YbPhYRQYAMPZtUv06CG+/LLcp8sbRNeuXYUQ77777oIFC+49g8PR0XH79u2rVq36+uuvDRwO/4s3MgMwkKVLxeTJIjpadOli6Cj/9eKLL1pYWDywpN7Z2Xnnzp1dunSpVavWm2++aahseADL6AHo3b3n865cKbZsEU2bGjpNZV2/fj0gIOCTTz4ZNWqUobNACGZgAPStuFiMGiX+/lscPixcXAyd5jHUr19/x44d3bp1s7OzCw0NNXQcUGAA9CknRwwaJGrXFnv26Pr5vLrQpEmTuLi4Xr162dnZ3XsTJgyIRRwA9OXKFeHvL/z8RHS0jO11T4sWLeLi4kaPHh0XF2foLNUdBQZAL44dE506iXfe0dvzeXWnZcuW69evHzFixP79+w2dpVpjEQcA3Vu/XowZI5YsESZ02W3Xrl1hYWGbN29u166dobNUUxQYAB2LihLffCNiY0WbNoaOomXbt29/9dVX4+LiWrVqZegs1REFBkBnVCoxYYLYvVts2SK8vAydRidiYmLeeeedXbt2+fj4GDpLtcMqRAC6UVgowsJEYaE4eNCwTzjUqcGDBxcWFgYFBe3du7dRo0aGjlO9UGAAdCAtTQwYIJo3F2vXGvwJh7o2fPjwkpKS7t2779u3z8tEJ5rGSe61QACM0V9/iQ4dRL9+YulSk2+ve1577bUJEyYEBgampqYaOks1wgwMgFbt3i2GDhUzZ4phwwwdRa/efffdvLy8oKCgPXv21K5d29BxqgUWcQDQnuXLxQcfiDVrRNeuho5iGJMnT962bdvu3budnJwMncX0UWAAtOH+83k3bxbVez3epEmTDh06tGPHjpo1axo6i4mjwAA8seJiMXq0OH9ebNwoXF0NncbA1Gr12LFjk5KSNm/ebGNjY+g4poxFHAAqQakUhw6VvysnR/TuLfLzxZ49tJcQQqFQzJ8/39XVdeDAgUql0tBxTBkFBqAS1q0THTuKjRsf3H71qujYUbRoIdatE7a2hkhmjMzMzFasWGFvbx8WFlZaWmroOCaLS4gAKqF/f7Fli3B2Flu3irZt/9n4xx9i4EDx4Yfif99fjHuKi4sHDx781FNPLVu2zOyh5xcnJydv3Ljx7NmzycnJOTk5Tk5OdevW9fHxCQ4O9vT0NEhg6VBgAB4lLU24u//ztaenOHBAeHmJbdvEyJFiwQIRHGzQcEatqKiob9++jRo1Wrx4sUKhuLdx8+bNX3zxReKlxOZ9mnu18XL0cLR7yu5O7p3c5NxrJ64lxiX6NPT5+OOPg/kX+ygUGIBH+fxzMWXKf79t1kyMHi1mzhTr1/93NgYNCgoKevXq1atXr4iIiEuXLo0aNepS9qWB0wc279vczLycD3HKVGV/bftrw5QNjRwbLVmypEmTJvrPLAsKDECFiotFo0YiOfl/NtrZiT//FPzdWjl5eXkJCQkqleqVV17pOrlrwNgAhZmi4kPUZep9C/bt+WLPypUrAwMD9ZNTOiziAFChlSsfbC8hRGGh+OILwa+/lePg4KBUKge9Mijsl7Cub3V9ZHsJIRRmiq5vdh2+ZviL4S/y6mdNmIEB0KykRPj4iKSk8vdOmyamTtVvIClduHChS5cuw1YP8+7i/bjHXj5yeengpXv27Hn22Wd1kU1qFBgAzebMEePHVzRg+XIRHq6vNFJSq9Xt2rXzecOn02udqnaGQ8sPnf7h9IkTJx5eyljN8a8DgAY5OeKLLx4x5tVXNd7gDCGEEKtXr842y+44smOVz+D/qn+BbcEvv/yixVSmgRkYAA1GjhTLlmnca2cnWrYUrVuLN9+s5g8/rFjTpk37zeun6eJhQWZZ9P8V/7W9tIa9ots4yx7jrcodlnQoKea1mL///vv+WnwIXqcCoHxbtpTfXq1aiQEDRGCgeO45YVX+37a47/Tp07dKbmlqr7JS8fNYZYfhlh1HWsYvLon+P6Wdk+L54eW8Qa2xf+Pb5rcTEhJatWql48gyocAAPOTGDTFmzP9sadhQvPaaCAsTjRoZKJOUNm7c6Bfsp2nvuV2lHUdY+Q4wF0J4B5inJJadWFdaboEJIfyC/WJjYymwf6PAAFNWUFAQFxcXHx+fnJycnp5ubm7u6enp6enZq1evgIAAy3Jfl5yVJQID/7t0vlMnMXmy6N1bsILg8SUkJDQe0ljTXo9m5k71/rkkqDATdX0VltYarxA+7f/0n8v+1H5EmfFfJGCaLl68OGTIENd6rtOXTU9vmO4a5toxsmP7z9rb9bO7aH9xzJQxbm5uEydOzM7O/p/Dbt8WgYHiwgUhhGjVSuzdK+LjRd++tFfVJCcnO3o4atp7v73uuXa8rGl3jZMKBw+HlJQUbYaTHzMwwNQolcr3339/6ZqlgRMDIxdEWte0fnhMnw/75Kbkbovc5uPjM23atLfeeksIIYqLRa9eIiFBCCE++UR8+qkod4qGSsvKyqpZp1KvtUy/WGbvqmg7ROPfybWca2VkZGgvmimgwACTkpGRMXjw4LvudyMSI2ydKnq/iaOHY+js0IzxGZGhkSdPnpz3ww9WQUHi8GHh4iLWrROdqnjTEv7NwcGhKK/okcPKVGLtf5SjVtYwM9c4piivyNFR42SueuKyAGA6cnNzu3Tp4tTd6fVVr1fcXve5PO0yad+kE1knhtSrV7Z/v3jpJXHlCu2lLe7u7rkpuY8c9vus4qD3rR64oviAnOScunXrai+aKSmuZNgAACAASURBVKDAABNx71mx9fvW7/9p/8e6W8jaznr06tHnmtf+NCRE/PYb76XUIm9v7+QzDz1J8n8dW13i6Wf2dMd/Jl+Ft8q/NzclMYUn0z+AAgNMRFRU1BXllUFfDarCseaW5qNXj553cv8hHquhVf379z+16VQFAxI2lKaeLatZ2+z6ybJrJ1Txi0suH1GVPzI2gTeEPYAncQCmIDc395lnnnnz9zfdfdwfPVqDoz8fPfvj2cOHD/O4B20pLS318PAYf2C8cyPnh/de3Kea3e9OqfK/WxzcFV9drmn20OKErKtZs56flZqaWv6dD9UVMzDAFMydO9e7v3cl2ys5sazc7c+FPXfzzs1du3ZpNVq1ZmFh8d5778V+GlvuXu8A8x9u1/qx5L//zLheTnsJITZO3fjuu+/SXg+gwABTsHbt2g4jOlQw4Pfvi8daFtz7Jy5SWe4YhZni+eHPx8TE6CZjNfXee++lHUxLOqThlTSVcOXolZt7b06cOFGLqUwDBQZI7/r165dTLzdqr/EhT2Wl4vIR1Xvbbe/9E/q9jaaRLYNbbty4kU8WtMjW1vbHH39c/MrinJs5VTg8NyV34dCFP/74o52dndazyY4CA6R3+vRpr7ZeFbzn93h0iXtT88YdzJt2N2/a3bxmHY0jnRs75xfnP/h4DjyZfv36fTjuw3kD5+Wl5j3Wgfnp+fMGzfvPm/8ZMGCAjrJJjQIDpJeSkuLoXtEtrr/PKtnyhXKCc8GCIUW5yY+YXTm4OyQnP2LlNx7X+++/P+alMTM6zbh+8nolD7n+5/XIjpGjQkZ9+OGHOs0mL57EAUjvkc8rejPG5uapsoTY0qOrSpIOFn6aYFfBJKyWc63MzEwdxKzuPv7442effXbMgDHPhDzTf0p/B3cHTSPz0/O3TN9ybv25RfMXDR48WJ8h5cIMDJCek5PTnZw7FQ3wNGvRz2L4QptJe22Li9SaFnHcU5hT6OTkpO2MEEKIgQMHnj9/voNThwi/iHmD5h1cejA5Mbkgs0AIcTvrdnJi8qFlh+YNnjet+bR2tdqdP3+e9qoYMzBAeu7u7nnbK/XhSoO25l3HWl8+WlrBmNyUXE9PTy1Fw4OcnJxmzJjx8ccfx8XFbdq0acOcDenp6RkZGc7Ozm5ubi1atJgcNrnPyj4ODhrnZ7iPAgOk98wzz9xIuFHJwR7NFcmJGq8f5qXmWZZYOjuXc9cttMje3n7IkCFDhgwxdBC5cQkRkJ6Pj09t69o3T9+szOCcm+pGz2t85vmpTaf69u1rxtu/IAP+MwVMQUhIyLHVx8rddf1k2a7ZxXdy1UKIwlvqv+NVPd7V+ECHP9b8ERISoquUgFbxLETAFKSkpHj7en9y/BMnzwfXX1w6oFr1zt3bWequb1nZOir8R1hY2ZZ/CTFxW+LOSTsTExMtLPhwARKgwAATMXny5D039oxcNrJqh5cUlXzV4at5X8xjBgZZUGCAibh9+3bHjh19XvPp9na3xz1WrVYvHbG0iaLJzz//rItsgC5woQAwEWVlZbGxsf7+/nZP2T039LnKH6guU8dMjhGXxOJ9i3UXD9A6FnEApuDIkSNNmzZVKpXbt2/fO21v7KexZary35nygKK8ovkvzi/9s3Tr1q02Nhof8gsYIQoMkN6GDRuCg4MXLVr0zDPPtGjR4ujRo2Uny6a3mn568+kKjlKVqPbO2zu1+dSuDbtu3769du3aegsMaAWfgQFymzNnzowZM2JjY9u0afPv7XFxcZMnT04pSGkZ0tKnp49TPSdHD8dSZWl+Rn76hfTEbYlntpzp3K5zZGRkixYtDBUeeBIUGCArtVodERERHR29detWLy+vcsecPn06NjZ2//79KSkpycnJ1tbWbm5uDRo0CAwMDA4Orl+/vp4zA1pEgQFSUiqVI0eOvHbtWmxsbJ06dQwdBzAAPgMD5JOTk9OrV6+SkpJdu3bRXqi2KDBAMlevXu3YsWPLli1//fVX1g2iOqPAAJmcPn06ICDg7bffjoqK4pG7qOa4kRmQxs6dO8PDwxcsWBAcHGzoLIDh8RscIIelS5eGh4evXbuW9gLuYQYGGLt7y+V//vnnffv2eXt7GzoOYCwoMMCoFRcXjxo16u+//z58+DAvSgb+jQIDjNft27dfeuklGxub3bt329raGjoOYFz4DAwwUikpKZ07d/b29l63bh3tBTyMAgOMUWJior+/f0hICMvlAU24hAgYnd27dw8dOjQqKio0NNTQWQDjxW92gHFZsWJFWFhYdHQ07QVUjBkYYESioqJmz569d+/epk2bGjoLYOwoMMAoqFSqd9555/jx44cOHXJ1dTV0HEACFBhgeIWFhUOGDCkpKdm9e3etWrUMHQeQA5+BAQaWlpYWEBDg6uq6ZcsW2guoPAoMMKRLly516dKlf//+S5YssbDgigjwGPg/DGAwhw8ffvHFF2fMmDFs2DBDZwHkwwwMMIx169aFhIT89NNPtBdQNczAAAOIior69ttvd+zY0bJlS0NnAWRFgQF6pVKp3nvvvb179x46dKhevXqGjgNIjAID9Ofu3buvvvpqdnb2gQMHHBwcDB0HkBufgQF6kp2dHRgYaGNjExcXR3sBT44CA/Th8uXL/v7+HTt2XLZsmaWlpaHjAKaAAgN07tixY506dZowYUJkZKRCoTB0HMBEUGCAFmRnZ2vatWHDhv79+y9atGjs2LH6jASYPAoMeFJlZWW+vr6ffvrpw7sWLVo0bty4uLi4fv366T8YYNpYhQg8qfj4+JSUlOnTp9evX//111+/t1GtVkdERERHRx84cMDLy8uwCQGTRIEBT2rFihX3vhg9enTdunX79OmjVCpHjhx57dq1ffv21alTx7DxAFPFJUTgieTn569fv/7+tyNGjNi7d29QUFBxcfGuXbtoL0B3KDDgiSxbtiwnJ+f+txkZGYGBgQ0bNvztt99sbGwMGAwweRQYUHWlpaWzZ89+eOPRo0fz8vIMEgmoPigwoOrWrFmTlJT08PZz584NHDhQqVTqPxJQfSjUarWhMwBSKi4ufvbZZy9duqRpwJAhQ1atWmVmxq+JgE7wfy2giubNm1dBewkhfv31148++khveYDqhgIDqiIlJeWrr76qeEybNm2srKz0kweohrgPDKiKsWPHZmRkPLy9devWvXv37tmzZ/v27W1tbfUfDKg+KDDgsS1dunTTpk3/3uLt7f3aa6+9/PLLDRs2NFQqoLphEQequwsXLly8eDElJSU3N7dOnTpubm6+vr4VvCv5999/DwwMvP9t3759J02a1LVrVx4zD+gZBYZqKiUlZc6cOevXr88syvT09XRwd7B1tL2ddTs/Pf/q8atNPJsMHjz47bffdnR0/PdRJ06c6NmzZ25urhCiX79+ERERbdq0MdCfAKjuKDBUO4WFhV999dWCBQtaDW/Vbmi7+q3qPzCgTFV2+cjlw8sPX9p66YMPPhg/fry5ubkQ4u+//27Xrl1eXp6Tk9P3338fHh5uiPgA/kGBoXq5ceNGcHCwlY/VoC8HOXk6VTw47XzamnfX1Leov2bNmpKSkubNm2dmZrZs2XLt2rWNGzfWT2AAmlBgqEbOnTsXGBjoP9G/x/gelTykTFUW82HM1c1Xb926devWrYkTJ3799df3JmQADIsCQ3Vx69at9u3bd/64c/th7R/32M2fbd721bY9e/Z06tRJF9kAVAEFhuqiX79+qmdVg74cVIVj1Wr14rDF/u7+Dz+6F4Ch8CQOVAvbt28/celEyGchVTtcoVC8Mv+VZb8uO3v2rHaDAagyZmCoFtq2bdvuw3YtB7Z85MgylVAohKK8X+12ztqpOKaIjo7Wfj4Aj48ZGEzf2bNnr2Zdrbi9lLdFXGTxho+LT0SXqkrKH9PljS5bdm7hRV+AkaDAYPrWr1/fMqSi9rpyTBXhV/hUPcXAL6zahVpYWJc/zNrOuknnJnFxcTpJCeAxUWAwfQcOHPDp6aNpb9r5su97FQW9b9n+FctHnqpZYLP4+HitpgNQRRQYTN/Nmzefqv+Upr0rx9x18lQEjKnUe0+eqvdUcnKy9qIBqDoKDKYvLS3N3tW+3F2p58qSDqnsXc2WDC+a2qxwRsc7SYdVFZzK3t0+JSVFNzEBPB4KDKbP0tJSVVJ+LV39QyWECJxoOWpljWmJds6NzeYGF93O0rg0t6y0zMKClxABRoECg+lzd3fPSy1/6WBRvloI0aSzhRBCYSb6fWJ1J1d9dofGSVjOzRxPT08d5QTwWCgwmL769etnJmWWu8u5oZkQ4m7BP1Mul6fNzMzF7WyNM7CsK1kVvCoMgD5RYDB9gYGBidsSy93l1cbcwkqRcvafKZeqVK0uE67eGl9NeWbLmaCgIJ2kBPCYKDCYvuDg4DNbz5QWlz68y95N0WWMxb4f/7l1+e94Vd0WZj49y/+UKz89P/tsdkBAgA6zAqg0HiWFamHAgAG2PWy7v9P94V2qErFpWnFRfpnrM2ZZSereH1jZu5U/A/tt4m/NFc2joqJ0HBZApVBgqBYSExM7BnaMSIyo4VCjamfIvJwZ1THq7Nmzzs7O2s0GoGq4hIhqoXnz5iOHjFw6Yqm6rCq/sRXfKV4UuujTTz+lvQDjwQwM1UVJSUlQUJC5n/kLX7+gUGhcpvGwstKyJcOXNLdrvmzZMp2lA/DYmIGhWjh58uTkyZOjo6PzD+UvfXVpyV0ND5x/SGF2YVTfKA+lx4IFC3SaEMDjosBg+nbs2NGnT59OnTrVrl173759TyuejuwQeXbHI15NqVarj60+9kX7LwY+N3DDhg3W1hqeUQ/AQLiECBO3cuXK//znP9HR0Z07d76/ccOGDZMnTy5zKWs3tJ1ffz97t/95UmLW1axTG08dWXmkrk3dr7/++t8HAjAeFBhMWVRUVFRU1NatW5s2bfrArtLS0tjY2PXr12/btq3UutTJw6mGQ43b2bfz0vJshW1wcPDgwYN79er1WJ+WAdAnCgymSaVSjR8/Pj4+fuvWrRU/vVClUqWnpycnJ+fl5dWpU8fNzc3V1ZXeAowfBQYTpFQqhw8fnp2dHRMT4+DgYOg4AHSCRRwwNTk5OYGBgdbW1nFxcbQXYMIoMJiUq1ev+vv7t2rVavny5VZWlXrJMgBJUWAwHWfOnOnSpcuoUaOioqLMzPhvGzBxvFsWJmL37t1Dhw6dM2fOyy+/bOgsAPSB31JhCn7++eewsLDo6GjaC6g+mIFBelFRUd9+++2OHTt8fX0NnQWA/lBgkJhKpXr33Xf37dt3+PDhim/2AmB6KDDISqlUhoeHZ2ZmHjhwgOXyQDXEZ2CQ0r2bvcrKyrZu3Up7AdUTBQb5JCcnd+3atVWrVr/++quNjY2h4wAwDAoMkklMTPT39x82bBg3ewHVHJ+BQSb3bvaKiooKDQ01dBYABkaBQRpr164dN25cdHR0ly5dDJ0FgOFRYJDDvZu9tm3b5ufnZ+gsAIwCBQZjp1arP/jgg61btx46dKhevXqGjgPAWFBgMGpKpfLVV1/NyMg4ePAgy+UB/BuLuGC8cnJyevXqVVpays1eAB5GgcFIpaSkdOvWrWXLlr/99hs3ewF4GAUGY5SYmNihQ4ewsDBu9gKgCX81wDBu3Ljxxx9/lLtrz549PXr0mDFjxvvvv6/nVAAkQoHBMGbOnNmtW7eHO2zdunVhYWG//fYbtyoDqJhCrVYbOgOqnbKyMi8vr5s3b7q7u8fHxzdu3Pje9ns3e23atKlly5aGTQjA+DEDgwHs3r375s2bQojU1NSgoKD09PR7N3stWrTo4MGDtBeAyuA+MBjAkiVL7n+dlJQ0YMAALy+vrKysAwcOODo6GjAYAIlwCRH6lpqa6uHh8cBGNze3ixcv1qpVyyCRAMiIS4jQt6ioqIc3pqWlTZw4Uf9hAMiLGRj0Kicnp0mTJtnZ2eXunTp16rRp0/SbCICsmIFBryIjIzW1lxAiIiJi8eLF+swDQF7MwKA/169f9/LyeuSww4cPP//883rIA0BqrEKE/rz99tsV7PX29u7UqVO7du2eeeYZvUUCIC8KDHqyatWqzZs3P7y9f//+ISEhffr0qVu3rv5TAZAXlxChD1evXn3++efT09Pvb2nZsuWYMWNCQ0O58QtA1TADw2NLTk7es2fPjRs30tLS1Gq1u7u7p6dnt27dPD09yx2flZXVq1ev++0VGBg4ZcqUzp076zEyABPEDAyVVVpa+tNPPy1ZsuTy5WM9e4oGDYS7u1AoRFqauHJF/P678PJqO2LEiDfeeMPS0vL+UXfu3OnSpcuJEyeEEO3atYuKiurQoYPh/hAATAcFhkrZunXrhAkTGjS4+P77IiBAWDw0dVepxIED4uuvxcWLT8+cOTM4OFgIUVpa2rNnz3379gkhPv7446lTp/672wDgSVBgeAS1Wv35558vWfLp4sWiZ89Hj9+zR4waJcLCPo6IiBg0aNCmTZvq1Kmzbt26Ll266D4sgGqEAsMjvPHGG3/9tSgmRri6VvaQzEzx4ovi2jWva9euBQYGrl271t7eXpcZAVRHFBgq8u23365a9Z/4eGFn93gH3r0runUTDg5B27Zt0000ANUdBQaNjh49Onjw80ePCg2rCx8hNVW0by9++WU/Cw4B6AIFBo26d+/+6qt7Xn216mdYs0bMmvXckSNHFAqF9nIBgBAUGDTZsWPHpElBf/4pzM0f3FVcbF637pSsrP9eVVQo1Kmp011dCx4YqVaLdu3Ep5/G3luUCABaxI3MKN/q1avHji2nvYQQsbHPjh59dODARGtrlRAiMdHthx/8H24vIYRCId58U6xevZoCA6B1zMBQDpVK5e7ufuJEZr165ew9ccKzTZub97+dNq2XrW3x++/vLfdUGRmiaVOntLQ0Kysr3YQFUE3xPjCU49KlS46O5beXEOLf7SWEWLeuxcCBf2k6lYuLqFcv59y5c9pNCAAUGMqRkpJSyZWHFy86l5UpvL0zKxjj6SlSUlK0kwwA/j8KDOVIT0+v5G3L69a1GDQoseIxbm4iNTVVC7EA4F8oMJSjRo0ad+5UamRMzKML7M4dYWtrq4VYAPAvFBjK4eHhUZkp07VrThkZNR/4SOxhycmCl1UC0DoKDOXw8vJKShKlpY8YFhPTYuDAR0y/ysrE33+Lhg0bai0cAAghKDCUy8XFpVGjtgcPPmJYZT4AO3ZMuLj4anrXJQBUGQWG8g0cOHDt2ooGpKXVunjRuXPnKxWfZ906wV3MAHSBG5lRvoyMjObNmx8/nlm//pOcRDRvLo4fv1b/Sc4CAOVhBobyubi4jBkzZsqUJzrJJ5+IESP+Q3sB0AVmYNCooKDgueeemzTp/KhRVTn8559FRMTTf/zxh6Ojo7ajAQAFhgpdvHixS5cuy5enBwU93oF79ojQUJc9e/Y0a9ZMN9EAVHdcQkT5SkpKxo4de/78+XXr1o0c6f7DD49x7MKFIizM7bfffqO9AOgOMzCUo7Cw8OWXXzY3N1+zZo2tre3Vq1dDQkI8PE7PmCF8fSs68K+/xIcfiitXnt2wYcPTTz+tr7wAqiNmYHhQenp6QECAq6trTEzMvUdANWjQ4Pjx4/37/xAU5Na/v1i8WCQn/88hqali6VIREiJ69HANDIw6efIk7QVA15iB4X9cvny5d+/egwcPjoyMfHjv7du3t2zZEhsbu3v37tzcdHd3oVCI1FRRq5Zz9+7dQ0JC+vfvX6tWLf3HBlANUWD4r2PHjg0cOHDatGlvvPHGIwcrlcrU1FS1Wu3u7m5jY6OHeADwbxQY/rFjx47w8PAlS5b069fP0FkA4NH4DAxCCLFs2bLw8PCYmBjaC4AsLAwdAIY3Y8aMhQsX7t+/39vb29BZAKCyKLBqTaVSvf322ydPnjx8+LCLi4uh4wDAY6DAqq/CwsLQ0FClUrlr1y6WDgKQDp+BVVPZ2dm9evWqXbv2li1baC8AMqLAqqMrV674+/t37tx56dKllpaWho4DAFVBgVU7x48f79ix44QJEyIjIxUKhaHjAEAV8RlY9fL777+HhYXNmzfvxRdfNHQWAHgizMCqkRUrVgwbNiwmJob2AmACKDCTcvbs2ZdeeqmoqOjhXVFRUREREfv27evUqZP+gwGA1nEJ0aT88ssva9euLS0tXbt2rbm5+b2NKpVq3LhxBw4c2L9/f926dQ2bEAC0hWchmg61Wv30009fvnxZCPHmm2/OmzdPCKFUKocNG5aTkxMTE2Nvb2/ojACgNVxCNB0HDhy4115CiPnz53/55Ze3bt3q2bNnjRo14uLiaC8AJoYZmOl45ZVXVq1a9e8t7u7uL7300qxZs8zM+E0FgKmhwExEamqqh4fHw9u3bdsWFBSk/zwAoGv8Ym4ivvvuu3K3Dx8+/MiRI3oOAwB6wAzMFGRmZvr4+GRnZ5e718PDIz4+vlGjRnpOBQA6xQzMFHz22Wea2ksIkZKS0rt378zMTH1GAgBdo8Ckd+7cuR9++KHiMX///ffq1av1kwcA9IMbmeVWVlb2+uuva9obEBDQvXv3gICANm3a1KxZU5/BAEDXKDC5zZ0799ChQw9sHDBgwJAhQ/r37+/g4GCQVACgByzikNiBAwc6d+58/1tPT8+xY8e+9tpr7u7uBkwFAPrBDMwoXLt2LTY2dtu2bVevXk1JSSkqKnJzc6tbt+7zzz8fEhLSqVOn+w82vO/s2bODBg2693W9evU++OCD119/3draWu/ZAcAwmIEZ2NmzZydPnrznyJ4WfVv49vd19XZ1cHewsrUqSC+4dfPWxb0XE2IT1JnqiIiI8PDw+w/USE5ObtOmTXp6uhBi0qRJn3/+OdUFoLqhwAxGpVJNnjx5wYoFQe8HBYwNsLDSOBu+cvRKzOSYmrdrRkdHN27cODc318/P7/r1666urqtWrerevbs+YwOAkaDADCM/Pz80NPRa6bXXf3nd1sm2MofsX7h/9/Tdy5cvHzdu3MWLF3v27LlmzZratWvrOioAGCcKzABKS0t79+5d3Kh46OyhZhaPcSvexX0XF4UtKswqnDlz5nvvvae7hABg/LiR2QDee++9DJuMsB/CHqu9hBDeAd7D5g9zcXEZMmSIjrIBgCyYgenbgQMHBg0f9MmJT2zsbap2ho1TNzpedfzll1+0GwwA5EKB6VunTp2ajm3aLrRdlc9QUlTy6bOf7ly/s02bNloMBgByocD0Kj4+PvTN0CknpyjMFA/vLcxWb/lCWcPerEylLshUv/iNtU2tcoYJIfbO26s4onjg9ZUAUK3wGZhebdiwoe3LbcttLyHEstfuenexGDDNKmS6dS0XxcapSk3nafNim23bthUXF+ssKQAYOwpMrzZv3uw3wE/T3ov7VKX/v7OcPM1Sz2qcHNdyqVW7We34+HitJwQAWVBg+qNSqS5fu+zeTOODChu0M4+LVOanqctKxZ/rS54fXtGDvjx9PS9cuKCDmAAgBwpMfzIyMuxq25mZa/x3PnyhzZ089Vf+hcteu9v7A6v2r1hWcDYHd4eUlBQdxAQAOVBg+pObm1vDvkYFA+o0VIT9YFOYLU6uKy3IeMTiGlsn25ycHK0GBACZUGD64+zsXJBZUMGAK0fLDi0r+eqynV+I+ZLhdw8tL6lgcH56vqurq7YzAoA0KDD9qV27dunt0lJlabl7VSVi3qCiLmMs7WorXv+5xnNhFtH/pyxTaTxbXkqeh4eHrrICgNGjwPRHoVD4+vpePnK53L3KQnVBZpmTp5kQQmEmeoy3KspTFxdqPFvS4aSWLVvqKCoAGD8KTK9CQkJObTxV7i5bR0WDtuZnd/4zP7udpW7Y3szGvvzzZCZlmuWb8SQOANUZBaZXgwcPPr72ePGd8m9AfuNXm8tHVBunKnfPKTkVq3pjtcbXrBz86eALL7ygUJR/QzQAVAc8Skrfhg4dWtSsqM/kPlU+Q25K7jetv0lISPD09NRiMACQCzMwffviiy92z9mdcSmjaoer1epfJ/w6duxY2gtANUeB6VujRo1mfz17bsjcO7l3qnD41i+32qTaTJkyRevBAEAuFJgBjBgxIrRP6A8DfshPz3+sA3d+t/PMsjMxMTHW1tY6ygYAsqDADGPWrFnhvcMjO0ZqWlX/gDu5d5aOXHot+lp8fLybm5uu4wGA8WMRhyFt2LBh/PjxLu1d+k3p59Gs/LuS7xbcPfjTwe3fbH9l8CszZ86sUaOih1EBQPVBgRlYUVHR7Nmz586dW2JX4tvP183Hzd7V3qamTW5ybk5KzoXdF5IOJ/UL7PfZZ5/5+PgYOiwAGBEKzCio1eqTJ09u27YtKSkpPT29oKCgbt267u7u/v7+vXv3rlmzpqEDAoDRocAAAFJiEQcAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUiutV0gAAAIdJREFUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBKFBgAQEoUGABAShQYAEBK/w/WA1DLn9IT0QAAAABJRU5ErkJggg==" /><!-- --></p>
<p>A plot of the connections between functions in the pipeline. The node labels are the <code>rmonad</code> ids. The red arrow indicates a ‘nest’ relationship. Green nodes are OK, orange nodes raised a warning, red nodes (none appear in this graph) represent errors.</p>
<p>If the pipeline fails, the last valid result is saved, for example</p>
<pre class="sourceCode r" id="cb38"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1"><span class="kw">read_gff</span>(gff<span class="op">$</span>not_a_gff1)</a></code></pre>
<pre><code>## 
## 
##     
##       Rmonad supports docstrings. If an block begins with a string, this
##       string is extracted and stored. Python has something similar, where the
##       first string in a function is cast as documentation.
##       
##       The `as_monad` function takes an expression and wraps its result into a
##       context. It also handles the extraction of this docstring. The result
##       here is used at more than one place in the pipeline. Rather than
##       accessing it later as a global, it will be funneled bach in.
##       
## 
## N1&gt; &quot;&quot;
##  * WARNING: number of columns of result is not a multiple of vector length (arg 1)
##  * WARNING: 3 parsing failures.
## row # A tibble: 3 x 5 col     row   col  expected    actual         file expected   &lt;int&gt; &lt;chr&gt;     &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt; actual 1     1  &lt;NA&gt; 9 columns 1 columns literal data file 2     2  &lt;NA&gt; 9 columns 1 columns literal data row 3     3  &lt;NA&gt; 9 columns 1 columns literal data
## 
## 
## 
##     
##     The %&gt;&gt;% operator applies the function described in this block to the
##     input on the left-hand-side. This corresponds to the UNIX '|' or magrittr's
##     '%&gt;%'. It differs from them in that it is a monadic bind operator, rather
##     than an application operator. It carries a context along with the
##     computations. The context can store past values, performance information,
##     this docstring, and links to the parent chunk. The context is a directed
##     graph of code chunks and their metadata.
##     
## 
## N2&gt; &quot;function (.) 
## {
##     for (col in c(&quot;seqid&quot;, &quot;type&quot;, &quot;start&quot;, &quot;stop&quot;)) {
##         if (any(is.na(.[[col]]))) {
##             stop(&quot;GFFError: Column '&quot;, col, &quot;' may not have missing values&quot;)
##         }
##     }
##     .
## }&quot;
##  * ERROR: GFFError: Column 'type' may not have missing values
## 
##  ----------------- 
## 
## # A tibble: 3 x 9
##                seqid source  type start  stop score strand phase  attr
##                &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;dbl&gt;  &lt;chr&gt; &lt;int&gt; &lt;chr&gt;
## 1   Once upon a time   &lt;NA&gt;  &lt;NA&gt;    NA    NA    NA   &lt;NA&gt;    NA  &lt;NA&gt;
## 2 there was a prince   &lt;NA&gt;  &lt;NA&gt;    NA    NA    NA   &lt;NA&gt;    NA  &lt;NA&gt;
## 3 who liked to plumb   &lt;NA&gt;  &lt;NA&gt;    NA    NA    NA   &lt;NA&gt;    NA  &lt;NA&gt;
##  *** FAILURE ***</code></pre>
<p>This makes debugging much simpler. We don’t need to jump back and rerun small parts of the pipeline. The failing object, and all intermediate data, could be saved. This could also allow for much richer bug reporting.</p>
<p>Overall, in <code>rmonad</code>, the output of a pipeline is not just the effluent of the last pipeline, but the collection of all the nodes along the way. The pipeline itself becomes data that can be computed upon.</p>
</section>
</section>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
